// @volcengine/openapi/lib/services/imagex/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const service_1 = __importDefault(require("../../base/service"));
/**
 * Class representing a ImagexService.
 */
class ImagexService extends service_1.default {
    /**
     * 创建 ImagexService
     * @param options 初始化配置
     */
    constructor(options) {
        super(Object.assign(Object.assign({}, options), { serviceName: "imagex" }));
        /**
         * @function UpdateImageDomainVolcOrigin
         */
        this.UpdateImageDomainVolcOrigin = this.createAPI("UpdateImageDomainVolcOrigin", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function DelDomain
         */
        this.DelDomain = this.createAPI("DelDomain", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function AddDomainV1
         */
        this.AddDomainV1 = this.createAPI("AddDomainV1", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageDomainIPAuth
         */
        this.UpdateImageDomainIPAuth = this.createAPI("UpdateImageDomainIPAuth", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateRefer
         */
        this.UpdateRefer = this.createAPI("UpdateRefer", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageDomainUaAccess
         */
        this.UpdateImageDomainUaAccess = this.createAPI("UpdateImageDomainUaAccess", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateHttps
         */
        this.UpdateHttps = this.createAPI("UpdateHttps", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageDomainDownloadSpeedLimit
         */
        this.UpdateImageDomainDownloadSpeedLimit = this.createAPI("UpdateImageDomainDownloadSpeedLimit", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateResponseHeader
         */
        this.UpdateResponseHeader = this.createAPI("UpdateResponseHeader", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageDomainAreaAccess
         */
        this.UpdateImageDomainAreaAccess = this.createAPI("UpdateImageDomainAreaAccess", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateDomainAdaptiveFmt
         */
        this.UpdateDomainAdaptiveFmt = this.createAPI("UpdateDomainAdaptiveFmt", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageDomainConfig
         */
        this.UpdateImageDomainConfig = this.createAPI("UpdateImageDomainConfig", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateAdvance
         */
        this.UpdateAdvance = this.createAPI("UpdateAdvance", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageDomainBandwidthLimit
         */
        this.UpdateImageDomainBandwidthLimit = this.createAPI("UpdateImageDomainBandwidthLimit", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateSlimConfig
         */
        this.UpdateSlimConfig = this.createAPI("UpdateSlimConfig", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function SetDefaultDomain
         */
        this.SetDefaultDomain = this.createAPI("SetDefaultDomain", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageVolcCdnAccessLog
         */
        this.DescribeImageVolcCdnAccessLog = this.createAPI("DescribeImageVolcCdnAccessLog", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function VerifyDomainOwner
         */
        this.VerifyDomainOwner = this.createAPI("VerifyDomainOwner", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetResponseHeaderValidateKeys
         */
        this.GetResponseHeaderValidateKeys = this.createAPI("GetResponseHeaderValidateKeys", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetDomainConfig
         */
        this.GetDomainConfig = this.createAPI("GetDomainConfig", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["DomainName", "ServiceId"],
        });
        /**
         * @function GetDomainOwnerVerifyContent
         */
        this.GetDomainOwnerVerifyContent = this.createAPI("GetDomainOwnerVerifyContent", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Domain"],
        });
        /**
         * @function GetServiceDomains
         */
        this.GetServiceDomains = this.createAPI("GetServiceDomains", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function DeleteImageMonitorRules
         */
        this.DeleteImageMonitorRules = this.createAPI("DeleteImageMonitorRules", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
        });
        /**
         * @function DeleteImageMonitorRecords
         */
        this.DeleteImageMonitorRecords = this.createAPI("DeleteImageMonitorRecords", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
        });
        /**
         * @function CreateImageMonitorRule
         */
        this.CreateImageMonitorRule = this.createAPI("CreateImageMonitorRule", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
        });
        /**
         * @function UpdateImageMonitorRule
         */
        this.UpdateImageMonitorRule = this.createAPI("UpdateImageMonitorRule", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
        });
        /**
         * @function UpdateImageMonitorRuleStatus
         */
        this.UpdateImageMonitorRuleStatus = this.createAPI("UpdateImageMonitorRuleStatus", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
        });
        /**
         * @function GetImageAlertRecords
         */
        this.GetImageAlertRecords = this.createAPI("GetImageAlertRecords", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
        });
        /**
         * @function GetImageMonitorRules
         */
        this.GetImageMonitorRules = this.createAPI("GetImageMonitorRules", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["Limit", "Offset", "AppId", "NamePtn", "RuleId"],
        });
        /**
         * @function CreateImageSettingRule
         */
        this.CreateImageSettingRule = this.createAPI("CreateImageSettingRule", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageSettingRule
         */
        this.DeleteImageSettingRule = this.createAPI("DeleteImageSettingRule", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageSettingRulePriority
         */
        this.UpdateImageSettingRulePriority = this.createAPI("UpdateImageSettingRulePriority", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageSettingRule
         */
        this.UpdateImageSettingRule = this.createAPI("UpdateImageSettingRule", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageSettings
         */
        this.GetImageSettings = this.createAPI("GetImageSettings", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["AppId", "Category"],
        });
        /**
         * @function GetImageSettingRuleHistory
         */
        this.GetImageSettingRuleHistory = this.createAPI("GetImageSettingRuleHistory", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["AppId", "SettingId", "Offset", "Limit"],
        });
        /**
         * @function GetImageSettingRules
         */
        this.GetImageSettingRules = this.createAPI("GetImageSettingRules", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["AppId", "SettingId"],
        });
        /**
         * @function CreateImageMigrateTask
         */
        this.CreateImageMigrateTask = this.createAPI("CreateImageMigrateTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageMigrateTask
         */
        this.DeleteImageMigrateTask = this.createAPI("DeleteImageMigrateTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Region", "TaskId"],
        });
        /**
         * @function ExportFailedMigrateTask
         */
        this.ExportFailedMigrateTask = this.createAPI("ExportFailedMigrateTask", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Region", "TaskId"],
        });
        /**
         * @function UpdateImageTaskStrategy
         */
        this.UpdateImageTaskStrategy = this.createAPI("UpdateImageTaskStrategy", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function TerminateImageMigrateTask
         */
        this.TerminateImageMigrateTask = this.createAPI("TerminateImageMigrateTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Region", "TaskId"],
        });
        /**
         * @function GetVendorBuckets
         */
        this.GetVendorBuckets = this.createAPI("GetVendorBuckets", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageMigrateTasks
         */
        this.GetImageMigrateTasks = this.createAPI("GetImageMigrateTasks", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Region", "TaskId", "ServiceId", "Offset", "Limit", "TaskNamePtn", "Status"],
        });
        /**
         * @function RerunImageMigrateTask
         */
        this.RerunImageMigrateTask = this.createAPI("RerunImageMigrateTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Region", "TaskId"],
        });
        /**
         * @function GetImageAddOnTag
         */
        this.GetImageAddOnTag = this.createAPI("GetImageAddOnTag", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Key", "Type"],
        });
        /**
         * @function DescribeImageXCubeUsage
         */
        this.DescribeImageXCubeUsage = this.createAPI("DescribeImageXCubeUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "StartTime", "EndTime", "Interval"],
        });
        /**
         * @function DescribeImageXSourceRequestBandwidth
         */
        this.DescribeImageXSourceRequestBandwidth = this.createAPI("DescribeImageXSourceRequestBandwidth", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "DomainNames",
                "Regions",
                "UserCountry",
                "UserProvince",
                "Protocols",
                "Isp",
                "GroupBy",
                "StartTime",
                "EndTime",
                "Interval",
            ],
        });
        /**
         * @function DescribeImageXSourceRequestTraffic
         */
        this.DescribeImageXSourceRequestTraffic = this.createAPI("DescribeImageXSourceRequestTraffic", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "DomainNames",
                "Regions",
                "UserCountry",
                "UserProvince",
                "Protocols",
                "Isp",
                "GroupBy",
                "StartTime",
                "EndTime",
                "Interval",
            ],
        });
        /**
         * @function DescribeImageXSourceRequest
         */
        this.DescribeImageXSourceRequest = this.createAPI("DescribeImageXSourceRequest", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "DomainNames",
                "Regions",
                "UserCountry",
                "UserProvince",
                "Protocols",
                "Isp",
                "DataTypes",
                "GroupBy",
                "StartTime",
                "EndTime",
                "Interval",
                "DetailedCode",
            ],
        });
        /**
         * @function DescribeImageXStorageUsage
         */
        this.DescribeImageXStorageUsage = this.createAPI("DescribeImageXStorageUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "BucketNames",
                "GroupBy",
                "StartTime",
                "EndTime",
                "Interval",
                "IsRetrieval",
            ],
        });
        /**
         * @function DescribeImageXBucketRetrievalUsage
         */
        this.DescribeImageXBucketRetrievalUsage = this.createAPI("DescribeImageXBucketRetrievalUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "BucketNames", "GroupBy", "StartTime", "EndTime", "IsRetrieval"],
        });
        /**
         * @function DescribeImageXAddOnQPSUsage
         */
        this.DescribeImageXAddOnQPSUsage = this.createAPI("DescribeImageXAddOnQPSUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["AdvFeat", "ServiceIds", "StartTime", "EndTime", "Interval"],
        });
        /**
         * @function DescribeImageXAIRequestCntUsage
         */
        this.DescribeImageXAIRequestCntUsage = this.createAPI("DescribeImageXAIRequestCntUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "AdvFeats",
                "Templates",
                "EnableBillingRate",
                "GroupBy",
                "StartTime",
                "EndTime",
                "Interval",
            ],
        });
        /**
         * @function DescribeImageXSummary
         */
        this.DescribeImageXSummary = this.createAPI("DescribeImageXSummary", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "Timestamp"],
        });
        /**
         * @function DescribeImageXDomainTrafficData
         */
        this.DescribeImageXDomainTrafficData = this.createAPI("DescribeImageXDomainTrafficData", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "DomainNames",
                "BillingRegion",
                "GroupBy",
                "StartTime",
                "EndTime",
                "Interval",
            ],
        });
        /**
         * @function DescribeImageXDomainBandwidthData
         */
        this.DescribeImageXDomainBandwidthData = this.createAPI("DescribeImageXDomainBandwidthData", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "DomainNames",
                "BillingRegion",
                "GroupBy",
                "StartTime",
                "EndTime",
                "Interval",
            ],
        });
        /**
         * @function DescribeImageXDomainBandwidthNinetyFiveData
         */
        this.DescribeImageXDomainBandwidthNinetyFiveData = this.createAPI("DescribeImageXDomainBandwidthNinetyFiveData", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "DomainNames", "BillingRegion", "StartTime", "EndTime"],
        });
        /**
         * @function DescribeImageXBucketUsage
         */
        this.DescribeImageXBucketUsage = this.createAPI("DescribeImageXBucketUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "BucketNames", "GroupBy", "StartTime", "EndTime"],
        });
        /**
         * @function DescribeImageXBillingRequestCntUsage
         */
        this.DescribeImageXBillingRequestCntUsage = this.createAPI("DescribeImageXBillingRequestCntUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "AdvFeats", "GroupBy", "StartTime", "EndTime", "Interval"],
        });
        /**
         * @function DescribeImageXRequestCntUsage
         */
        this.DescribeImageXRequestCntUsage = this.createAPI("DescribeImageXRequestCntUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "AdvFeats",
                "Templates",
                "GroupBy",
                "StartTime",
                "EndTime",
                "Interval",
            ],
        });
        /**
         * @function DescribeImageXBaseOpUsage
         */
        this.DescribeImageXBaseOpUsage = this.createAPI("DescribeImageXBaseOpUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "GroupBy", "StartTime", "EndTime", "Interval"],
        });
        /**
         * @function DescribeImageXCompressUsage
         */
        this.DescribeImageXCompressUsage = this.createAPI("DescribeImageXCompressUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "GroupBy", "StartTime", "EndTime", "Interval"],
        });
        /**
         * @function DescribeImageXScreenshotUsage
         */
        this.DescribeImageXScreenshotUsage = this.createAPI("DescribeImageXScreenshotUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "StartTime", "EndTime", "Interval"],
        });
        /**
         * @function DescribeImageXVideoClipDurationUsage
         */
        this.DescribeImageXVideoClipDurationUsage = this.createAPI("DescribeImageXVideoClipDurationUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "StartTime", "EndTime", "Interval"],
        });
        /**
         * @function DescribeImageXMultiCompressUsage
         */
        this.DescribeImageXMultiCompressUsage = this.createAPI("DescribeImageXMultiCompressUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "StartTime", "EndTime", "Interval"],
        });
        /**
         * @function DescribeImageXEdgeRequest
         */
        this.DescribeImageXEdgeRequest = this.createAPI("DescribeImageXEdgeRequest", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "DomainNames",
                "Regions",
                "UserCountry",
                "UserProvince",
                "Protocols",
                "Isp",
                "DataTypes",
                "GroupBy",
                "StartTime",
                "EndTime",
                "Interval",
                "DetailedCode",
            ],
        });
        /**
         * @function DescribeImageXEdgeRequestBandwidth
         */
        this.DescribeImageXEdgeRequestBandwidth = this.createAPI("DescribeImageXEdgeRequestBandwidth", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "DomainNames",
                "Regions",
                "UserCountry",
                "UserProvince",
                "Protocols",
                "Isp",
                "GroupBy",
                "StartTime",
                "EndTime",
                "Interval",
            ],
        });
        /**
         * @function DescribeImageXEdgeRequestTraffic
         */
        this.DescribeImageXEdgeRequestTraffic = this.createAPI("DescribeImageXEdgeRequestTraffic", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "DomainNames",
                "Regions",
                "UserCountry",
                "UserProvince",
                "Protocols",
                "Isp",
                "GroupBy",
                "StartTime",
                "EndTime",
                "Interval",
            ],
        });
        /**
         * @function DescribeImageXEdgeRequestRegions
         */
        this.DescribeImageXEdgeRequestRegions = this.createAPI("DescribeImageXEdgeRequestRegions", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "StartTime", "EndTime"],
        });
        /**
         * @function DescribeImageXMirrorRequestHttpCodeByTime
         */
        this.DescribeImageXMirrorRequestHttpCodeByTime = this.createAPI("DescribeImageXMirrorRequestHttpCodeByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXMirrorRequestHttpCodeOverview
         */
        this.DescribeImageXMirrorRequestHttpCodeOverview = this.createAPI("DescribeImageXMirrorRequestHttpCodeOverview", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXMirrorRequestTraffic
         */
        this.DescribeImageXMirrorRequestTraffic = this.createAPI("DescribeImageXMirrorRequestTraffic", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXMirrorRequestBandwidth
         */
        this.DescribeImageXMirrorRequestBandwidth = this.createAPI("DescribeImageXMirrorRequestBandwidth", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXServerQPSUsage
         */
        this.DescribeImageXServerQPSUsage = this.createAPI("DescribeImageXServerQPSUsage", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "StartTime", "EndTime", "Interval"],
        });
        /**
         * @function DescribeImageXHitRateTrafficData
         */
        this.DescribeImageXHitRateTrafficData = this.createAPI("DescribeImageXHitRateTrafficData", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "DomainNames", "GroupBy", "StartTime", "EndTime", "Interval"],
        });
        /**
         * @function DescribeImageXHitRateRequestData
         */
        this.DescribeImageXHitRateRequestData = this.createAPI("DescribeImageXHitRateRequestData", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceIds", "DomainNames", "GroupBy", "StartTime", "EndTime", "Interval"],
        });
        /**
         * @function DescribeImageXCDNTopRequestData
         */
        this.DescribeImageXCDNTopRequestData = this.createAPI("DescribeImageXCDNTopRequestData", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceIds",
                "DomainNames",
                "IPVersion",
                "Country",
                "KeyType",
                "ValueType",
                "StartTime",
                "EndTime",
                "Limit",
                "Offset",
            ],
        });
        /**
         * @function DescribeImageXHeifEncodeFileInSizeByTime
         */
        this.DescribeImageXHeifEncodeFileInSizeByTime = this.createAPI("DescribeImageXHeifEncodeFileInSizeByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXHeifEncodeFileOutSizeByTime
         */
        this.DescribeImageXHeifEncodeFileOutSizeByTime = this.createAPI("DescribeImageXHeifEncodeFileOutSizeByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXHeifEncodeSuccessCountByTime
         */
        this.DescribeImageXHeifEncodeSuccessCountByTime = this.createAPI("DescribeImageXHeifEncodeSuccessCountByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXHeifEncodeSuccessRateByTime
         */
        this.DescribeImageXHeifEncodeSuccessRateByTime = this.createAPI("DescribeImageXHeifEncodeSuccessRateByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXHeifEncodeDurationByTime
         */
        this.DescribeImageXHeifEncodeDurationByTime = this.createAPI("DescribeImageXHeifEncodeDurationByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXHeifEncodeErrorCodeByTime
         */
        this.DescribeImageXHeifEncodeErrorCodeByTime = this.createAPI("DescribeImageXHeifEncodeErrorCodeByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXExceedResolutionRatioAll
         */
        this.DescribeImageXExceedResolutionRatioAll = this.createAPI("DescribeImageXExceedResolutionRatioAll", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXExceedFileSize
         */
        this.DescribeImageXExceedFileSize = this.createAPI("DescribeImageXExceedFileSize", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXExceedCountByTime
         */
        this.DescribeImageXExceedCountByTime = this.createAPI("DescribeImageXExceedCountByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXServiceQuality
         */
        this.DescribeImageXServiceQuality = this.createAPI("DescribeImageXServiceQuality", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Region"],
        });
        /**
         * @function GetImageXQueryApps
         */
        this.GetImageXQueryApps = this.createAPI("GetImageXQueryApps", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Source", "Project"],
        });
        /**
         * @function GetImageXQueryRegions
         */
        this.GetImageXQueryRegions = this.createAPI("GetImageXQueryRegions", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Source", "Appid", "OS"],
        });
        /**
         * @function GetImageXQueryDims
         */
        this.GetImageXQueryDims = this.createAPI("GetImageXQueryDims", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Source", "Appid", "OS"],
        });
        /**
         * @function GetImageXQueryVals
         */
        this.GetImageXQueryVals = this.createAPI("GetImageXQueryVals", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Dim", "Source", "Appid", "OS", "Keyword"],
        });
        /**
         * @function DescribeImageXUploadCountByTime
         */
        this.DescribeImageXUploadCountByTime = this.createAPI("DescribeImageXUploadCountByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXUploadDuration
         */
        this.DescribeImageXUploadDuration = this.createAPI("DescribeImageXUploadDuration", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXUploadSuccessRateByTime
         */
        this.DescribeImageXUploadSuccessRateByTime = this.createAPI("DescribeImageXUploadSuccessRateByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXUploadFileSize
         */
        this.DescribeImageXUploadFileSize = this.createAPI("DescribeImageXUploadFileSize", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXUploadErrorCodeByTime
         */
        this.DescribeImageXUploadErrorCodeByTime = this.createAPI("DescribeImageXUploadErrorCodeByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXUploadErrorCodeAll
         */
        this.DescribeImageXUploadErrorCodeAll = this.createAPI("DescribeImageXUploadErrorCodeAll", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXUploadSpeed
         */
        this.DescribeImageXUploadSpeed = this.createAPI("DescribeImageXUploadSpeed", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXUploadSegmentSpeedByTime
         */
        this.DescribeImageXUploadSegmentSpeedByTime = this.createAPI("DescribeImageXUploadSegmentSpeedByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXCdnSuccessRateByTime
         */
        this.DescribeImageXCdnSuccessRateByTime = this.createAPI("DescribeImageXCdnSuccessRateByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXCdnSuccessRateAll
         */
        this.DescribeImageXCdnSuccessRateAll = this.createAPI("DescribeImageXCdnSuccessRateAll", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXCdnErrorCodeByTime
         */
        this.DescribeImageXCdnErrorCodeByTime = this.createAPI("DescribeImageXCdnErrorCodeByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXCdnErrorCodeAll
         */
        this.DescribeImageXCdnErrorCodeAll = this.createAPI("DescribeImageXCdnErrorCodeAll", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXCdnDurationDetailByTime
         */
        this.DescribeImageXCdnDurationDetailByTime = this.createAPI("DescribeImageXCdnDurationDetailByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXCdnDurationAll
         */
        this.DescribeImageXCdnDurationAll = this.createAPI("DescribeImageXCdnDurationAll", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXCdnReuseRateByTime
         */
        this.DescribeImageXCdnReuseRateByTime = this.createAPI("DescribeImageXCdnReuseRateByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXCdnReuseRateAll
         */
        this.DescribeImageXCdnReuseRateAll = this.createAPI("DescribeImageXCdnReuseRateAll", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXCdnProtocolRateByTime
         */
        this.DescribeImageXCdnProtocolRateByTime = this.createAPI("DescribeImageXCdnProtocolRateByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientFailureRate
         */
        this.DescribeImageXClientFailureRate = this.createAPI("DescribeImageXClientFailureRate", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientDecodeSuccessRateByTime
         */
        this.DescribeImageXClientDecodeSuccessRateByTime = this.createAPI("DescribeImageXClientDecodeSuccessRateByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientDecodeDurationByTime
         */
        this.DescribeImageXClientDecodeDurationByTime = this.createAPI("DescribeImageXClientDecodeDurationByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientQueueDurationByTime
         */
        this.DescribeImageXClientQueueDurationByTime = this.createAPI("DescribeImageXClientQueueDurationByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientErrorCodeByTime
         */
        this.DescribeImageXClientErrorCodeByTime = this.createAPI("DescribeImageXClientErrorCodeByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientErrorCodeAll
         */
        this.DescribeImageXClientErrorCodeAll = this.createAPI("DescribeImageXClientErrorCodeAll", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientLoadDuration
         */
        this.DescribeImageXClientLoadDuration = this.createAPI("DescribeImageXClientLoadDuration", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientLoadDurationAll
         */
        this.DescribeImageXClientLoadDurationAll = this.createAPI("DescribeImageXClientLoadDurationAll", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientSdkVerByTime
         */
        this.DescribeImageXClientSdkVerByTime = this.createAPI("DescribeImageXClientSdkVerByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientFileSize
         */
        this.DescribeImageXClientFileSize = this.createAPI("DescribeImageXClientFileSize", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientTopFileSize
         */
        this.DescribeImageXClientTopFileSize = this.createAPI("DescribeImageXClientTopFileSize", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientCountByTime
         */
        this.DescribeImageXClientCountByTime = this.createAPI("DescribeImageXClientCountByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientQualityRateByTime
         */
        this.DescribeImageXClientQualityRateByTime = this.createAPI("DescribeImageXClientQualityRateByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientTopQualityURL
         */
        this.DescribeImageXClientTopQualityURL = this.createAPI("DescribeImageXClientTopQualityURL", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientDemotionRateByTime
         */
        this.DescribeImageXClientDemotionRateByTime = this.createAPI("DescribeImageXClientDemotionRateByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientTopDemotionURL
         */
        this.DescribeImageXClientTopDemotionURL = this.createAPI("DescribeImageXClientTopDemotionURL", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXClientScoreByTime
         */
        this.DescribeImageXClientScoreByTime = this.createAPI("DescribeImageXClientScoreByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXSensibleCountByTime
         */
        this.DescribeImageXSensibleCountByTime = this.createAPI("DescribeImageXSensibleCountByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXSensibleCacheHitRateByTime
         */
        this.DescribeImageXSensibleCacheHitRateByTime = this.createAPI("DescribeImageXSensibleCacheHitRateByTime", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXSensibleTopSizeURL
         */
        this.DescribeImageXSensibleTopSizeURL = this.createAPI("DescribeImageXSensibleTopSizeURL", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXSensibleTopResolutionURL
         */
        this.DescribeImageXSensibleTopResolutionURL = this.createAPI("DescribeImageXSensibleTopResolutionURL", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXSensibleTopRamURL
         */
        this.DescribeImageXSensibleTopRamURL = this.createAPI("DescribeImageXSensibleTopRamURL", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DescribeImageXSensibleTopUnknownURL
         */
        this.DescribeImageXSensibleTopUnknownURL = this.createAPI("DescribeImageXSensibleTopUnknownURL", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function CreateBatchProcessTask
         */
        this.CreateBatchProcessTask = this.createAPI("CreateBatchProcessTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetBatchProcessResult
         */
        this.GetBatchProcessResult = this.createAPI("GetBatchProcessResult", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetBatchTaskInfo
         */
        this.GetBatchTaskInfo = this.createAPI("GetBatchTaskInfo", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["TaskId", "ServiceId"],
        });
        /**
         * @function AIProcess
         */
        this.AIProcess = this.createAPI("AIProcess", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function CreateImageAITask
         */
        this.CreateImageAITask = this.createAPI("CreateImageAITask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function CreateImageAIProcessQueue
         */
        this.CreateImageAIProcessQueue = this.createAPI("CreateImageAIProcessQueue", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageAIProcessQueue
         */
        this.DeleteImageAIProcessQueue = this.createAPI("DeleteImageAIProcessQueue", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function CreateImageAIProcessCallback
         */
        this.CreateImageAIProcessCallback = this.createAPI("CreateImageAIProcessCallback", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageAIProcessQueue
         */
        this.UpdateImageAIProcessQueue = this.createAPI("UpdateImageAIProcessQueue", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageAIProcessQueueStatus
         */
        this.UpdateImageAIProcessQueueStatus = this.createAPI("UpdateImageAIProcessQueueStatus", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageAIProcessDetail
         */
        this.DeleteImageAIProcessDetail = this.createAPI("DeleteImageAIProcessDetail", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageAIProcessQueues
         */
        this.GetImageAIProcessQueues = this.createAPI("GetImageAIProcessQueues", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["SearchPtn", "Limit", "Offset"],
        });
        /**
         * @function GetImageAITasks
         */
        this.GetImageAITasks = this.createAPI("GetImageAITasks", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceId",
                "QueueId",
                "TaskId",
                "Marker",
                "Limit",
                "Status",
                "StartTime",
                "EndTime",
                "SearchPtn",
            ],
        });
        /**
         * @function GetImageAIDetails
         */
        this.GetImageAIDetails = this.createAPI("GetImageAIDetails", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "QueueId",
                "TaskId",
                "StartTime",
                "EndTime",
                "Status",
                "SearchPtn",
                "Limit",
                "Offset",
                "ServiceId",
            ],
        });
        /**
         * @function UpdateImageResourceStatus
         */
        this.UpdateImageResourceStatus = this.createAPI("UpdateImageResourceStatus", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateFileStorageClass
         */
        this.UpdateFileStorageClass = this.createAPI("UpdateFileStorageClass", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetImageStorageFiles
         */
        this.GetImageStorageFiles = this.createAPI("GetImageStorageFiles", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId", "Marker", "Limit", "Prefix", "Delimiter"],
        });
        /**
         * @function DeleteImageUploadFiles
         */
        this.DeleteImageUploadFiles = this.createAPI("DeleteImageUploadFiles", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function CreateFileRestore
         */
        this.CreateFileRestore = this.createAPI("CreateFileRestore", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageUploadFiles
         */
        this.UpdateImageUploadFiles = this.createAPI("UpdateImageUploadFiles", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function CommitImageUpload
         */
        this.CommitImageUpload = this.createAPI("CommitImageUpload", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId", "SkipMeta"],
        });
        /**
         * @function UpdateImageFileCT
         */
        this.UpdateImageFileCT = this.createAPI("UpdateImageFileCT", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function ApplyVpcUploadInfo
         */
        this.ApplyVpcUploadInfo = this.createAPI("ApplyVpcUploadInfo", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "ServiceId",
                "StoreKey",
                "Prefix",
                "FileExtension",
                "ContentType",
                "StorageClass",
                "FileSize",
                "PartSize",
                "Overwrite",
            ],
        });
        /**
         * @function ApplyImageUpload
         */
        this.ApplyImageUpload = this.createAPI("ApplyImageUpload", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: [
                "ServiceId",
                "SessionKey",
                "UploadNum",
                "StoreKeys",
                "Prefix",
                "FileExtension",
                "Overwrite",
            ],
        });
        /**
         * @function GetImageUploadFile
         */
        this.GetImageUploadFile = this.createAPI("GetImageUploadFile", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId", "StoreUri"],
        });
        /**
         * @function GetImageUploadFiles
         */
        this.GetImageUploadFiles = this.createAPI("GetImageUploadFiles", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId", "Limit", "Marker"],
        });
        /**
         * @function GetImageUpdateFiles
         */
        this.GetImageUpdateFiles = this.createAPI("GetImageUpdateFiles", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId", "Type", "UrlPattern", "Offset", "Limit"],
        });
        /**
         * @function PreviewImageUploadFile
         */
        this.PreviewImageUploadFile = this.createAPI("PreviewImageUploadFile", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId", "StoreUri"],
        });
        /**
         * @function GetImageEraseResult
         */
        this.GetImageEraseResult = this.createAPI("GetImageEraseResult", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageService
         */
        this.GetImageService = this.createAPI("GetImageService", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetAllImageServices
         */
        this.GetAllImageServices = this.createAPI("GetAllImageServices", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["SearchPtn"],
        });
        /**
         * @function CreateImageCompressTask
         */
        this.CreateImageCompressTask = this.createAPI("CreateImageCompressTask", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function FetchImageUrl
         */
        this.FetchImageUrl = this.createAPI("FetchImageUrl", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageStorageTTL
         */
        this.UpdateImageStorageTTL = this.createAPI("UpdateImageStorageTTL", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetCompressTaskInfo
         */
        this.GetCompressTaskInfo = this.createAPI("GetCompressTaskInfo", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["TaskId", "ServiceId"],
        });
        /**
         * @function GetUrlFetchTask
         */
        this.GetUrlFetchTask = this.createAPI("GetUrlFetchTask", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["Id", "ServiceId"],
        });
        /**
         * @function GetResourceURL
         */
        this.GetResourceURL = this.createAPI("GetResourceURL", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId", "Domain", "URI", "Tpl", "Proto", "Format", "Timestamp"],
        });
        /**
         * @function CreateImageFromUri
         */
        this.CreateImageFromUri = this.createAPI("CreateImageFromUri", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageFileKey
         */
        this.UpdateImageFileKey = this.createAPI("UpdateImageFileKey", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function CreateImageContentTask
         */
        this.CreateImageContentTask = this.createAPI("CreateImageContentTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageContentTaskDetail
         */
        this.GetImageContentTaskDetail = this.createAPI("GetImageContentTaskDetail", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageContentBlockList
         */
        this.GetImageContentBlockList = this.createAPI("GetImageContentBlockList", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function CreateImageTranscodeQueue
         */
        this.CreateImageTranscodeQueue = this.createAPI("CreateImageTranscodeQueue", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageTranscodeQueue
         */
        this.DeleteImageTranscodeQueue = this.createAPI("DeleteImageTranscodeQueue", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageTranscodeQueue
         */
        this.UpdateImageTranscodeQueue = this.createAPI("UpdateImageTranscodeQueue", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageTranscodeQueueStatus
         */
        this.UpdateImageTranscodeQueueStatus = this.createAPI("UpdateImageTranscodeQueueStatus", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageTranscodeQueues
         */
        this.GetImageTranscodeQueues = this.createAPI("GetImageTranscodeQueues", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Region", "SearchPtn", "Limit", "Offset"],
        });
        /**
         * @function CreateImageTranscodeTask
         */
        this.CreateImageTranscodeTask = this.createAPI("CreateImageTranscodeTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageTranscodeDetails
         */
        this.GetImageTranscodeDetails = this.createAPI("GetImageTranscodeDetails", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: [
                "QueueId",
                "TaskId",
                "Region",
                "StartTime",
                "EndTime",
                "Status",
                "SearchPtn",
                "Limit",
                "Offset",
            ],
        });
        /**
         * @function CreateImageTranscodeCallback
         */
        this.CreateImageTranscodeCallback = this.createAPI("CreateImageTranscodeCallback", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageTranscodeDetail
         */
        this.DeleteImageTranscodeDetail = this.createAPI("DeleteImageTranscodeDetail", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImagePSDetection
         */
        this.GetImagePSDetection = this.createAPI("GetImagePSDetection", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetImageSuperResolutionResult
         */
        this.GetImageSuperResolutionResult = this.createAPI("GetImageSuperResolutionResult", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetDenoisingImage
         */
        this.GetDenoisingImage = this.createAPI("GetDenoisingImage", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetImageDuplicateDetection
         */
        this.GetImageDuplicateDetection = this.createAPI("GetImageDuplicateDetection", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetImageOCRV2
         */
        this.GetImageOCRV2 = this.createAPI("GetImageOCRV2", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetImageBgFillResult
         */
        this.GetImageBgFillResult = this.createAPI("GetImageBgFillResult", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetSegmentImage
         */
        this.GetSegmentImage = this.createAPI("GetSegmentImage", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetImageSmartCropResult
         */
        this.GetImageSmartCropResult = this.createAPI("GetImageSmartCropResult", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageComicResult
         */
        this.GetImageComicResult = this.createAPI("GetImageComicResult", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageEnhanceResult
         */
        this.GetImageEnhanceResult = this.createAPI("GetImageEnhanceResult", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
        });
        /**
         * @function GetImageQuality
         */
        this.GetImageQuality = this.createAPI("GetImageQuality", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetLicensePlateDetection
         */
        this.GetLicensePlateDetection = this.createAPI("GetLicensePlateDetection", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetPrivateImageType
         */
        this.GetPrivateImageType = this.createAPI("GetPrivateImageType", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function CreateCVImageGenerateTask
         */
        this.CreateCVImageGenerateTask = this.createAPI("CreateCVImageGenerateTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function CreateHiddenWatermarkImage
         */
        this.CreateHiddenWatermarkImage = this.createAPI("CreateHiddenWatermarkImage", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function CreateHmExtractTask
         */
        this.CreateHmExtractTask = this.createAPI("CreateHmExtractTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageExifData
         */
        this.UpdateImageExifData = this.createAPI("UpdateImageExifData", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetImageDetectResult
         */
        this.GetImageDetectResult = this.createAPI("GetImageDetectResult", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetCVImageGenerateResult
         */
        this.GetCVImageGenerateResult = this.createAPI("GetCVImageGenerateResult", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function CreateImageHmExtract
         */
        this.CreateImageHmExtract = this.createAPI("CreateImageHmExtract", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId", "StoreUri", "ImageUrl", "Algorithm"],
        });
        /**
         * @function GetCVTextGenerateImage
         */
        this.GetCVTextGenerateImage = this.createAPI("GetCVTextGenerateImage", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetCVImageGenerateTask
         */
        this.GetCVImageGenerateTask = this.createAPI("GetCVImageGenerateTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function CreateImageHmEmbed
         */
        this.CreateImageHmEmbed = this.createAPI("CreateImageHmEmbed", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetCVAnimeGenerateImage
         */
        this.GetCVAnimeGenerateImage = this.createAPI("GetCVAnimeGenerateImage", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetComprehensiveEnhanceImage
         */
        this.GetComprehensiveEnhanceImage = this.createAPI("GetComprehensiveEnhanceImage", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageAiGenerateTask
         */
        this.GetImageAiGenerateTask = this.createAPI("GetImageAiGenerateTask", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["TaskId", "ServiceId"],
        });
        /**
         * @function GetImageEraseModels
         */
        this.GetImageEraseModels = this.createAPI("GetImageEraseModels", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Type"],
        });
        /**
         * @function GetDedupTaskStatus
         */
        this.GetDedupTaskStatus = this.createAPI("GetDedupTaskStatus", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["TaskId"],
        });
        /**
         * @function GetImageHmExtractTaskInfo
         */
        this.GetImageHmExtractTaskInfo = this.createAPI("GetImageHmExtractTaskInfo", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function CreateImageService
         */
        this.CreateImageService = this.createAPI("CreateImageService", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageService
         */
        this.DeleteImageService = this.createAPI("DeleteImageService", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageAuthKey
         */
        this.UpdateImageAuthKey = this.createAPI("UpdateImageAuthKey", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateResEventRule
         */
        this.UpdateResEventRule = this.createAPI("UpdateResEventRule", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateServiceName
         */
        this.UpdateServiceName = this.createAPI("UpdateServiceName", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateStorageRules
         */
        this.UpdateStorageRules = this.createAPI("UpdateStorageRules", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateStorageRulesV2
         */
        this.UpdateStorageRulesV2 = this.createAPI("UpdateStorageRulesV2", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageObjectAccess
         */
        this.UpdateImageObjectAccess = this.createAPI("UpdateImageObjectAccess", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageUploadOverwrite
         */
        this.UpdateImageUploadOverwrite = this.createAPI("UpdateImageUploadOverwrite", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageMirrorConf
         */
        this.UpdateImageMirrorConf = this.createAPI("UpdateImageMirrorConf", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetImageServiceSubscription
         */
        this.GetImageServiceSubscription = this.createAPI("GetImageServiceSubscription", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["AddOnType", "AddOnId", "AddOnKey"],
        });
        /**
         * @function GetImageAuthKey
         */
        this.GetImageAuthKey = this.createAPI("GetImageAuthKey", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function CreateImageAnalyzeTask
         */
        this.CreateImageAnalyzeTask = this.createAPI("CreateImageAnalyzeTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageAnalyzeTaskRun
         */
        this.DeleteImageAnalyzeTaskRun = this.createAPI("DeleteImageAnalyzeTaskRun", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageAnalyzeTask
         */
        this.DeleteImageAnalyzeTask = this.createAPI("DeleteImageAnalyzeTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageAnalyzeTaskStatus
         */
        this.UpdateImageAnalyzeTaskStatus = this.createAPI("UpdateImageAnalyzeTaskStatus", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageAnalyzeTask
         */
        this.UpdateImageAnalyzeTask = this.createAPI("UpdateImageAnalyzeTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageAnalyzeTasks
         */
        this.GetImageAnalyzeTasks = this.createAPI("GetImageAnalyzeTasks", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Region", "SearchPtn", "Limit", "Offset"],
        });
        /**
         * @function GetImageAnalyzeResult
         */
        this.GetImageAnalyzeResult = this.createAPI("GetImageAnalyzeResult", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["TaskId", "StartTime", "EndTime", "RunId", "Limit", "Offset", "File"],
        });
        /**
         * @function DeleteImageElements
         */
        this.DeleteImageElements = this.createAPI("DeleteImageElements", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageBackgroundColors
         */
        this.DeleteImageBackgroundColors = this.createAPI("DeleteImageBackgroundColors", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageStyle
         */
        this.DeleteImageStyle = this.createAPI("DeleteImageStyle", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function CreateImageStyle
         */
        this.CreateImageStyle = this.createAPI("CreateImageStyle", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageStyleMeta
         */
        this.UpdateImageStyleMeta = this.createAPI("UpdateImageStyleMeta", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function AddImageElements
         */
        this.AddImageElements = this.createAPI("AddImageElements", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function AddImageBackgroundColors
         */
        this.AddImageBackgroundColors = this.createAPI("AddImageBackgroundColors", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageStyle
         */
        this.UpdateImageStyle = this.createAPI("UpdateImageStyle", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
        });
        /**
         * @function GetImageFonts
         */
        this.GetImageFonts = this.createAPI("GetImageFonts", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageElements
         */
        this.GetImageElements = this.createAPI("GetImageElements", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Type", "SearchPtn", "Limit", "Offset"],
        });
        /**
         * @function GetImageBackgroundColors
         */
        this.GetImageBackgroundColors = this.createAPI("GetImageBackgroundColors", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageStyles
         */
        this.GetImageStyles = this.createAPI("GetImageStyles", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Type", "SearchPtn", "Limit", "Offset"],
        });
        /**
         * @function GetImageStyleDetail
         */
        this.GetImageStyleDetail = this.createAPI("GetImageStyleDetail", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["StyleId"],
        });
        /**
         * @function GetImageStyleResult
         */
        this.GetImageStyleResult = this.createAPI("GetImageStyleResult", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function DownloadCert
         */
        this.DownloadCert = this.createAPI("DownloadCert", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["CertID"],
        });
        /**
         * @function GetImageAllDomainCert
         */
        this.GetImageAllDomainCert = this.createAPI("GetImageAllDomainCert", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetCertInfo
         */
        this.GetCertInfo = this.createAPI("GetCertInfo", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["CertID"],
        });
        /**
         * @function GetAllCerts
         */
        this.GetAllCerts = this.createAPI("GetAllCerts", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function CreateImageTemplate
         */
        this.CreateImageTemplate = this.createAPI("CreateImageTemplate", {
            method: "POST",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function DeleteTemplatesFromBin
         */
        this.DeleteTemplatesFromBin = this.createAPI("DeleteTemplatesFromBin", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function DeleteImageTemplate
         */
        this.DeleteImageTemplate = this.createAPI("DeleteImageTemplate", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function CreateImageTemplatesByImport
         */
        this.CreateImageTemplatesByImport = this.createAPI("CreateImageTemplatesByImport", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function CreateTemplatesFromBin
         */
        this.CreateTemplatesFromBin = this.createAPI("CreateTemplatesFromBin", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function GetImageTemplate
         */
        this.GetImageTemplate = this.createAPI("GetImageTemplate", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId", "TemplateName"],
        });
        /**
         * @function GetTemplatesFromBin
         */
        this.GetTemplatesFromBin = this.createAPI("GetTemplatesFromBin", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId", "TemplateNamePattern", "Offset", "Limit", "Asc"],
        });
        /**
         * @function GetAllImageTemplates
         */
        this.GetAllImageTemplates = this.createAPI("GetAllImageTemplates", {
            method: "GET",
            contentType: "json",
            Version: "2018-08-01",
            queryKeys: ["ServiceId", "TemplateNamePattern", "Offset", "Limit", "Asc"],
        });
        /**
         * @function CreateImageAuditTask
         */
        this.CreateImageAuditTask = this.createAPI("CreateImageAuditTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function CreateVideoAuditTask
         */
        this.CreateVideoAuditTask = this.createAPI("CreateVideoAuditTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function CreateAudioAuditTask
         */
        this.CreateAudioAuditTask = this.createAPI("CreateAudioAuditTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function DeleteImageAuditResult
         */
        this.DeleteImageAuditResult = this.createAPI("DeleteImageAuditResult", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetSyncAuditResult
         */
        this.GetSyncAuditResult = this.createAPI("GetSyncAuditResult", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function SingleImageAudit
         */
        this.SingleImageAudit = this.createAPI("SingleImageAudit", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function BatchImageAudit
         */
        this.BatchImageAudit = this.createAPI("BatchImageAudit", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["ServiceId"],
        });
        /**
         * @function UpdateImageAuditTaskStatus
         */
        this.UpdateImageAuditTaskStatus = this.createAPI("UpdateImageAuditTaskStatus", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateImageAuditTask
         */
        this.UpdateImageAuditTask = this.createAPI("UpdateImageAuditTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateAuditImageStatus
         */
        this.UpdateAuditImageStatus = this.createAPI("UpdateAuditImageStatus", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateVideoAuditTask
         */
        this.UpdateVideoAuditTask = this.createAPI("UpdateVideoAuditTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function UpdateAudioAuditTask
         */
        this.UpdateAudioAuditTask = this.createAPI("UpdateAudioAuditTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
        /**
         * @function GetImageAuditTasks
         */
        this.GetImageAuditTasks = this.createAPI("GetImageAuditTasks", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["Region", "Type", "AuditAbility", "Status", "TaskType", "Limit", "Offset"],
        });
        /**
         * @function GetImageAuditTaskResult
         */
        this.GetImageAuditTaskResult = this.createAPI("GetImageAuditTaskResult", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["AuditSuggestion", "ImageType", "Limit", "Marker", "Problem", "TaskId", "Type"],
        });
        /**
         * @function GetImageAuditResult
         */
        this.GetImageAuditResult = this.createAPI("GetImageAuditResult", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["TaskId", "Type", "Problem", "ImageType", "AuditSuggestion", "Limit", "Marker"],
        });
        /**
         * @function GetAuditEntrysCount
         */
        this.GetAuditEntrysCount = this.createAPI("GetAuditEntrysCount", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["TaskId"],
        });
        /**
         * @function GetVideoAuditResult
         */
        this.GetVideoAuditResult = this.createAPI("GetVideoAuditResult", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["AuditSuggestion", "ImageType", "Limit", "Marker", "Problem", "TaskId", "Type"],
        });
        /**
         * @function GetAudioAuditResult
         */
        this.GetAudioAuditResult = this.createAPI("GetAudioAuditResult", {
            method: "GET",
            contentType: "json",
            Version: "2023-05-01",
            queryKeys: ["AuditSuggestion", "ImageType", "Limit", "Marker", "Problem", "TaskId", "Type"],
        });
        /**
         * @function CreateImageRetryAuditTask
         */
        this.CreateImageRetryAuditTask = this.createAPI("CreateImageRetryAuditTask", {
            method: "POST",
            contentType: "json",
            Version: "2023-05-01",
        });
    }
}
exports.default = ImagexService;
