// @volcengine/openapi/lib/services/imagex/type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostDeleteImageElementsTypeEnum = exports.PostCreateImageStyleUnitEnum = exports.PostAddImageElementsTypeEnum = exports.GetImageTranscodeQueuesResTypeEnum = exports.GetImageTranscodeQueuesResStatusEnum = exports.GetImageStylesResUnitEnum = exports.GetGetImageTranscodeQueuesTypeEnum = exports.GetGetImageTranscodeQueuesStatusEnum = exports.GetGetImageStylesUnitEnum = exports.DeleteImageElementsBodyTypeEnum = exports.CreateImageStyleBodyUnitEnum = exports.AddImageElementsBodyTypeEnum = void 0;
/**
 * 要素类型，取值如下所示：
 * * image：图片要素；
 * * background：背景要素；
 * * mask：蒙版要素。
 * 取值image表示图片要素，background表示背景要素
 * @example "background"
 */
var AddImageElementsBodyTypeEnum;
(function (AddImageElementsBodyTypeEnum) {
    AddImageElementsBodyTypeEnum["Image"] = "image";
    AddImageElementsBodyTypeEnum["Background"] = "background";
})(AddImageElementsBodyTypeEnum = exports.AddImageElementsBodyTypeEnum || (exports.AddImageElementsBodyTypeEnum = {}));
/**
 * 尺寸单位。即样式画布宽度/高度的像素单位，取值为`px`。
 * 当前仅支持取值px表示像素
 * @default "px"
 * @example "px"
 */
var CreateImageStyleBodyUnitEnum;
(function (CreateImageStyleBodyUnitEnum) {
    CreateImageStyleBodyUnitEnum["Px"] = "px";
})(CreateImageStyleBodyUnitEnum = exports.CreateImageStyleBodyUnitEnum || (exports.CreateImageStyleBodyUnitEnum = {}));
/**
 * 要素类型，取值如下所示：
 * * image：图片要素；
 * * background：背景要素；
 * * mask：蒙版要素。
 * 取值image表示图片要素，background表示背景要素
 * @example "background"
 */
var DeleteImageElementsBodyTypeEnum;
(function (DeleteImageElementsBodyTypeEnum) {
    DeleteImageElementsBodyTypeEnum["Image"] = "image";
    DeleteImageElementsBodyTypeEnum["Background"] = "background";
})(DeleteImageElementsBodyTypeEnum = exports.DeleteImageElementsBodyTypeEnum || (exports.DeleteImageElementsBodyTypeEnum = {}));
/**
 * 尺寸单位。当前仅支持取值px表示像素。
 * @example "px"
 */
var GetGetImageStylesUnitEnum;
(function (GetGetImageStylesUnitEnum) {
    GetGetImageStylesUnitEnum["Px"] = "px";
})(GetGetImageStylesUnitEnum = exports.GetGetImageStylesUnitEnum || (exports.GetGetImageStylesUnitEnum = {}));
/**
 * 队列状态。取值如下所示：
 * - `Pending`：排队中
 * - `Running`：执行中
 * @example "Pending"
 */
var GetGetImageTranscodeQueuesStatusEnum;
(function (GetGetImageTranscodeQueuesStatusEnum) {
    GetGetImageTranscodeQueuesStatusEnum["Pending"] = "Pending";
    GetGetImageTranscodeQueuesStatusEnum["Running"] = "Running";
})(GetGetImageTranscodeQueuesStatusEnum = exports.GetGetImageTranscodeQueuesStatusEnum || (exports.GetGetImageTranscodeQueuesStatusEnum = {}));
/**
 * 队列类型。取值如下所示：
 * - `default`：表示账号默认队列，每个账号一个
 * - `user`：表示用户创建队列，个数将有配额限制
 * @example "default"
 */
var GetGetImageTranscodeQueuesTypeEnum;
(function (GetGetImageTranscodeQueuesTypeEnum) {
    GetGetImageTranscodeQueuesTypeEnum["Default"] = "default";
    GetGetImageTranscodeQueuesTypeEnum["User"] = "user";
})(GetGetImageTranscodeQueuesTypeEnum = exports.GetGetImageTranscodeQueuesTypeEnum || (exports.GetGetImageTranscodeQueuesTypeEnum = {}));
/**
 * 尺寸单位。当前仅支持取值px表示像素。
 * @example "px"
 */
var GetImageStylesResUnitEnum;
(function (GetImageStylesResUnitEnum) {
    GetImageStylesResUnitEnum["Px"] = "px";
})(GetImageStylesResUnitEnum = exports.GetImageStylesResUnitEnum || (exports.GetImageStylesResUnitEnum = {}));
/**
 * 队列状态。取值如下所示：
 * - `Pending`：排队中
 * - `Running`：执行中
 * @example "Pending"
 */
var GetImageTranscodeQueuesResStatusEnum;
(function (GetImageTranscodeQueuesResStatusEnum) {
    GetImageTranscodeQueuesResStatusEnum["Pending"] = "Pending";
    GetImageTranscodeQueuesResStatusEnum["Running"] = "Running";
})(GetImageTranscodeQueuesResStatusEnum = exports.GetImageTranscodeQueuesResStatusEnum || (exports.GetImageTranscodeQueuesResStatusEnum = {}));
/**
 * 队列类型。取值如下所示：
 * - `default`：表示账号默认队列，每个账号一个
 * - `user`：表示用户创建队列，个数将有配额限制
 * @example "default"
 */
var GetImageTranscodeQueuesResTypeEnum;
(function (GetImageTranscodeQueuesResTypeEnum) {
    GetImageTranscodeQueuesResTypeEnum["Default"] = "default";
    GetImageTranscodeQueuesResTypeEnum["User"] = "user";
})(GetImageTranscodeQueuesResTypeEnum = exports.GetImageTranscodeQueuesResTypeEnum || (exports.GetImageTranscodeQueuesResTypeEnum = {}));
/**
 * 要素类型，取值如下所示：
 * * image：图片要素；
 * * background：背景要素；
 * * mask：蒙版要素。
 * 取值image表示图片要素，background表示背景要素
 * @example "background"
 */
var PostAddImageElementsTypeEnum;
(function (PostAddImageElementsTypeEnum) {
    PostAddImageElementsTypeEnum["Image"] = "image";
    PostAddImageElementsTypeEnum["Background"] = "background";
})(PostAddImageElementsTypeEnum = exports.PostAddImageElementsTypeEnum || (exports.PostAddImageElementsTypeEnum = {}));
/**
 * 尺寸单位。即样式画布宽度/高度的像素单位，取值为`px`。
 * 当前仅支持取值px表示像素
 * @default "px"
 * @example "px"
 */
var PostCreateImageStyleUnitEnum;
(function (PostCreateImageStyleUnitEnum) {
    PostCreateImageStyleUnitEnum["Px"] = "px";
})(PostCreateImageStyleUnitEnum = exports.PostCreateImageStyleUnitEnum || (exports.PostCreateImageStyleUnitEnum = {}));
/**
 * 要素类型，取值如下所示：
 * * image：图片要素；
 * * background：背景要素；
 * * mask：蒙版要素。
 * 取值image表示图片要素，background表示背景要素
 * @example "background"
 */
var PostDeleteImageElementsTypeEnum;
(function (PostDeleteImageElementsTypeEnum) {
    PostDeleteImageElementsTypeEnum["Image"] = "image";
    PostDeleteImageElementsTypeEnum["Background"] = "background";
})(PostDeleteImageElementsTypeEnum = exports.PostDeleteImageElementsTypeEnum || (exports.PostDeleteImageElementsTypeEnum = {}));
