// @volcengine/openapi/lib/services/billing/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultService = exports.BillingService = void 0;
const service_1 = __importDefault(require("../../base/service"));
class BillingService extends service_1.default {
    constructor(options) {
        super(Object.assign(Object.assign({}, options), { host: "billing.volcengineapi.com", defaultVersion: "2022-01-01", serviceName: "billing" }));
        this.ListBill = this.createAPI("ListBill", {
            method: "POST",
            contentType: "json",
            Version: "2022-01-01",
        });
        this.ListBillDetail = this.createAPI("ListBillDetail", {
            method: "POST",
            contentType: "json",
            Version: "2022-01-01",
        });
        this.ListBillOverviewByProd = this.createAPI("ListBillOverviewByProd", {
            method: "POST",
            contentType: "json",
            Version: "2022-01-01",
        });
    }
}
exports.BillingService = BillingService;
exports.defaultService = new BillingService();
