// @volcengine/openapi/lib/services/sms/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultService = exports.SmsService = void 0;
const service_1 = __importDefault(require("../../base/service"));
var HttpMethod;
(function (HttpMethod) {
    HttpMethod["GET"] = "GET";
    HttpMethod["DELETE"] = "DELETE";
    HttpMethod["POST"] = "POST";
    HttpMethod["PUT"] = "PUT";
    HttpMethod["Patch"] = "PATCH";
    HttpMethod["Connect"] = "CONNECT";
    HttpMethod["Options"] = "OPTIONS";
    HttpMethod["Trace"] = "TRACE";
})(HttpMethod || (HttpMethod = {}));
const ServiceVersion20200101 = "2020-01-01";
const ServiceVersion20210101 = "2021-01-01";
const ServiceVersion20210111 = "2021-01-11";
class SmsService extends service_1.default {
    constructor(options) {
        super(Object.assign(Object.assign({}, options), { host: "sms.volcengineapi.com", serviceName: "volcSMS" }));
        /** sms group */
        this.Send = this.createAPI("SendSms", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20200101,
        });
        this.BatchSend = this.createAPI("SendBatchSms", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20210101,
        });
        this.Conversion = this.createAPI("Conversion", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20200101,
        });
        this.SendVerifyCode = this.createAPI("SendSmsVerifyCode", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20200101,
        });
        this.CheckVerifyCode = this.createAPI("CheckSmsVerifyCode", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20200101,
        });
        /* sms - end */
        /* vms */
        this.ApplyVmsTemplate = this.createAPI("ApplyVmsTemplate", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        this.GetVmsTemplateStatus = this.createAPI("GetVmsTemplateStatus", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        this.SendVms = this.createAPI("SendSms", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        /* vms -end */
        /* template -start */
        this.GetSmsTemplateAndOrderList = this.createAPI("GetSmsTemplateAndOrderList", {
            method: HttpMethod.GET,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        this.ApplySmsTemplate = this.createAPI("ApplySmsTemplate", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        this.DeleteSmsTemplate = this.createAPI("DeleteSmsTemplate", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        /* template end*/
        /* subAccount start*/
        this.GetSubAccountList = this.createAPI("GetSubAccountList", {
            method: HttpMethod.GET,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        this.GetSubAccountDetail = this.createAPI("GetSubAccountDetail", {
            method: HttpMethod.GET,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        this.InsertSmsSubAccount = this.createAPI("InsertSmsSubAccount", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        /* subAccount end*/
        /* signature start*/
        this.GetSignatureAndOrderList = this.createAPI("GetSignatureAndOrderList", {
            method: HttpMethod.GET,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        this.ApplySmsSignature = this.createAPI("ApplySmsSignature", {
            method: HttpMethod.GET,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        this.DeleteSignature = this.createAPI("DeleteSignature", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        this.UpdateSmsSignature = this.createAPI("UpdateSmsSignature", {
            method: HttpMethod.PUT,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        /* signature end*/
        /*batch import signatures*/
        this.ApplySignatureIdent = this.createAPI("ApplySignatureIdent", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        this.GetSignatureIdentList = this.createAPI("GetSignatureIdentList", {
            method: HttpMethod.GET,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
        this.BatchBindSignatureIdent = this.createAPI("BatchBindSignatureIdent", {
            method: HttpMethod.POST,
            contentType: "json",
            Version: ServiceVersion20210111,
        });
    }
}
exports.SmsService = SmsService;
exports.defaultService = new SmsService();
