// @volcengine/openapi/lib/services/sms/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* vms */
var SmsOrderStatus;
(function (SmsOrderStatus) {
    SmsOrderStatus[SmsOrderStatus["SmsOrder_REVIEWING"] = 1] = "SmsOrder_REVIEWING";
    SmsOrderStatus[SmsOrderStatus["SmsOrder_REJECTED"] = 2] = "SmsOrder_REJECTED";
    SmsOrderStatus[SmsOrderStatus["SmsOrder_PASSED"] = 3] = "SmsOrder_PASSED";
    SmsOrderStatus[SmsOrderStatus["SmsOrder_CLOSE"] = 4] = "SmsOrder_CLOSE";
    SmsOrderStatus[SmsOrderStatus["SmsOrder_EXEMPTED"] = 5] = "SmsOrder_EXEMPTED";
})(SmsOrderStatus || (SmsOrderStatus = {}));
var SmsChannelType;
(function (SmsChannelType) {
    SmsChannelType["SmsChannelTypeCnOTP"] = "CN_OTP";
    SmsChannelType["SmsChannelTypeCnNTC"] = "CN_NTC";
    SmsChannelType["SmsChannelTypeCnMKT"] = "CN_MKT";
    SmsChannelType["SmsChannelTypeCnVms"] = "CN_VMS";
    SmsChannelType["SmsChannelTypeI18nOTP"] = "I18N_OTP";
    SmsChannelType["SmsChannelTypeI18nMKT"] = "I18N_MKT";
})(SmsChannelType || (SmsChannelType = {}));
/* vms -end*/
/* template */
var Area;
(function (Area) {
    Area["AreaCN"] = "cn";
    Area["AreaOverseas"] = "overseas";
    Area["AreaAll"] = "all";
})(Area || (Area = {}));
/**
 * @name 签名更新接口参数
 */
var PurposeType;
(function (PurposeType) {
    PurposeType[PurposeType["self"] = 1] = "self";
    PurposeType[PurposeType["other"] = 2] = "other";
})(PurposeType || (PurposeType = {}));
