// @volcengine/openapi/lib/services/rtc/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultService = exports.RtcService = void 0;
const service_1 = __importDefault(require("../../base/service"));
class RtcService extends service_1.default {
    constructor(options) {
        super(Object.assign(Object.assign({}, options), { defaultVersion: "2022-06-01", serviceName: "rtc" }));
        this.StartRecord = this.createAPI("StartRecord", {
            method: "POST",
            contentType: "json",
            Version: "2022-06-01",
        });
        this.StopRecord = this.createAPI("StopRecord", {
            method: "POST",
            contentType: "json",
            Version: "2022-06-01",
        });
        this.GetRecordTask = this.createAPI("GetRecordTask", {
            method: "GET",
            contentType: "json",
            Version: "2022-06-01",
            queryKeys: ["AppId", "RoomId", "TaskId"],
        });
    }
}
exports.RtcService = RtcService;
exports.defaultService = new RtcService();
