// @volcengine/openapi/lib/services/livesaas/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LivesaasService = void 0;
const service_1 = __importDefault(require("../../base/service"));
const const_1 = require("./const");
class LivesaasService extends service_1.default {
    constructor(options) {
        super(Object.assign({ host: const_1.Domain[(options === null || options === void 0 ? void 0 : options.region) || "ap-singapore-1"], serviceName: "livesaas", defaultVersion: "2019-10-10" }, options));
        this.GetLivesaasUploadUserToken = () => {
            const policy = {
                Statement: [
                    {
                        Effect: "Allow",
                        Action: [...const_1.LivesaasUploadAction],
                        Resource: ["*"],
                    },
                ],
            };
            const UserToken = this.signSts2(policy);
            return UserToken;
        };
    }
}
exports.LivesaasService = LivesaasService;
