// @volcengine/openapi/lib/services/vikingdb/index/backend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Quant = exports.Distance = exports.IndexType = exports.ShardPolicy = void 0;
var ShardPolicy;
(function (ShardPolicy) {
    ShardPolicy["Custom"] = "custom";
    ShardPolicy["Auto"] = "auto";
})(ShardPolicy = exports.ShardPolicy || (exports.ShardPolicy = {}));
var IndexType;
(function (IndexType) {
    IndexType["Hnsw"] = "hnsw";
    IndexType["HnswHybrid"] = "hnsw_hybrid";
    IndexType["Flat"] = "flat";
    IndexType["Ivf"] = "ivf";
    IndexType["Diskann"] = "diskann";
})(IndexType = exports.IndexType || (exports.IndexType = {}));
var Distance;
(function (Distance) {
    Distance["Ip"] = "ip";
    Distance["l2"] = "l2";
    Distance["Cosine"] = "cosine";
})(Distance = exports.Distance || (exports.Distance = {}));
var Quant;
(function (Quant) {
    Quant["Int8"] = "int8";
    Quant["Float"] = "float";
    Quant["Fix16"] = "fix16";
    Quant["PQ"] = "pq";
})(Quant = exports.Quant || (exports.Quant = {}));
