// @volcengine/openapi/lib/services/vikingdb/index/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListIndexesResponse = exports.GetIndexInfoResponse = exports.ShardPolicy = exports.Quant = exports.IndexType = exports.Distance = void 0;
/* DropIndex start */
const types_1 = require("../types");
var backend_1 = require("./backend");
Object.defineProperty(exports, "Distance", { enumerable: true, get: function () { return backend_1.Distance; } });
Object.defineProperty(exports, "IndexType", { enumerable: true, get: function () { return backend_1.IndexType; } });
Object.defineProperty(exports, "Quant", { enumerable: true, get: function () { return backend_1.Quant; } });
Object.defineProperty(exports, "ShardPolicy", { enumerable: true, get: function () { return backend_1.ShardPolicy; } });
class GetIndexInfoResponse extends types_1.VikingdbResponse {
    constructor(IndexInfo, OriginalRequest, LogId) {
        super(OriginalRequest, LogId);
        this.IndexInfo = IndexInfo;
    }
}
exports.GetIndexInfoResponse = GetIndexInfoResponse;
class ListIndexesResponse extends types_1.VikingdbResponse {
    constructor(ListIndexes, OriginalRequest, LogId) {
        super(OriginalRequest, LogId);
        this.ListIndexes = ListIndexes;
    }
}
exports.ListIndexesResponse = ListIndexesResponse;
/* UpdateIndex end */
