// @volcengine/openapi/lib/services/live/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiveCasterService = void 0;
const service_1 = __importDefault(require("../../base/service"));
const const_1 = require("./const");
class LiveCasterService extends service_1.default {
    constructor(options) {
        super(Object.assign({ host: "live.volcengineapi.com", serviceName: "live", defaultVersion: "2018-01-01" }, options));
        this.GetCasterPlayerUserToken = () => {
            const policy = {
                Statement: [
                    {
                        Effect: "Allow",
                        Action: [...const_1.CasterPlayerAction],
                        Resource: ["*"],
                    },
                ],
            };
            const UserToken = this.signSts2(policy);
            return UserToken;
        };
    }
}
exports.LiveCasterService = LiveCasterService;
