// @volcengine/openapi/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.vikingdb = exports.maas = exports.sms = exports.cdn = exports.sts = exports.billing = exports.rocketmq = exports.livesaasOpenapi = exports.liveOpenapi = exports.tlsOpenapi = exports.vodOpenapi = exports.rtcOpenapi = exports.acep = exports.i18nOpenapi = exports.edit = exports.imagex = exports.iam = exports.Signer = exports.Service = void 0;
const service_1 = __importDefault(require("./base/service"));
exports.Service = service_1.default;
const sign_1 = __importDefault(require("./base/sign"));
exports.Signer = sign_1.default;
const iam = __importStar(require("./services/iam"));
exports.iam = iam;
const imagex = __importStar(require("./services/imagex"));
exports.imagex = imagex;
const edit = __importStar(require("./services/edit"));
exports.edit = edit;
const i18nOpenapi = __importStar(require("./services/i18nOpenapi"));
exports.i18nOpenapi = i18nOpenapi;
const acep = __importStar(require("./services/acep"));
exports.acep = acep;
const rtcOpenapi = __importStar(require("./services/rtc"));
exports.rtcOpenapi = rtcOpenapi;
const vodOpenapi = __importStar(require("./services/vod"));
exports.vodOpenapi = vodOpenapi;
const tlsOpenapi = __importStar(require("./services/tls"));
exports.tlsOpenapi = tlsOpenapi;
const liveOpenapi = __importStar(require("./services/live"));
exports.liveOpenapi = liveOpenapi;
const livesaasOpenapi = __importStar(require("./services/livesaas"));
exports.livesaasOpenapi = livesaasOpenapi;
const rocketmq = __importStar(require("./services/rocketmq"));
exports.rocketmq = rocketmq;
const billing = __importStar(require("./services/billing"));
exports.billing = billing;
const sts = __importStar(require("./services/sts"));
exports.sts = sts;
const cdn = __importStar(require("./services/cdn"));
exports.cdn = cdn;
const sms = __importStar(require("./services/sms"));
exports.sms = sms;
const maas = __importStar(require("./services/maas"));
exports.maas = maas;
const vikingdb = __importStar(require("./services/vikingdb"));
exports.vikingdb = vikingdb;
