// @volcengine/openapi/lib/services/vikingdb/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FieldType = exports.VikingdbRequestError = exports.VikingdbError = exports.VikingdbErrorCode = exports.VikingdbResponse = exports.task = exports.search = exports.embedding = exports.data = exports.index = exports.collection = exports.VikingdbService = void 0;
const collection_1 = require("./collection");
const index_1 = require("./index/index");
const data_1 = require("./data");
const embedding_1 = require("./embedding");
const search_1 = require("./search");
const custom_1 = require("./custom");
const task_1 = require("./task");
/**
 * @example
 * import { vikingdb } from '@volcengine/volc-sdk-nodejs'
 * const service = new vikingdb.VikingdbService({ ak, sk, region })
 * await service.collection.ListCollections()
 */
class VikingdbService {
    constructor({ ak, sk, region = "cn-beijing", sessionToken }) {
        this.collection = new collection_1.CollectionService(ak, sk, region, sessionToken);
        this.index = new index_1.IndexService(ak, sk, region, sessionToken);
        this.data = new data_1.DataService(ak, sk, region, sessionToken);
        this.embedding = new embedding_1.EmbeddingService(ak, sk, region, sessionToken);
        this.search = new search_1.SearchService(ak, sk, region, sessionToken);
        this.custom = new custom_1.CustomService(ak, sk, region, sessionToken);
        this.task = new task_1.TaskService(ak, sk, region, sessionToken);
    }
}
exports.VikingdbService = VikingdbService;
var collection_2 = require("./collection");
Object.defineProperty(exports, "collection", { enumerable: true, get: function () { return collection_2.collection; } });
var index_2 = require("./index/index");
Object.defineProperty(exports, "index", { enumerable: true, get: function () { return index_2.index; } });
var data_2 = require("./data");
Object.defineProperty(exports, "data", { enumerable: true, get: function () { return data_2.data; } });
var embedding_2 = require("./embedding");
Object.defineProperty(exports, "embedding", { enumerable: true, get: function () { return embedding_2.embedding; } });
var search_2 = require("./search");
Object.defineProperty(exports, "search", { enumerable: true, get: function () { return search_2.search; } });
var task_2 = require("./task");
Object.defineProperty(exports, "task", { enumerable: true, get: function () { return task_2.task; } });
var types_1 = require("./types");
Object.defineProperty(exports, "VikingdbResponse", { enumerable: true, get: function () { return types_1.VikingdbResponse; } });
Object.defineProperty(exports, "VikingdbErrorCode", { enumerable: true, get: function () { return types_1.VikingdbErrorCode; } });
Object.defineProperty(exports, "VikingdbError", { enumerable: true, get: function () { return types_1.VikingdbError; } });
Object.defineProperty(exports, "VikingdbRequestError", { enumerable: true, get: function () { return types_1.VikingdbRequestError; } });
Object.defineProperty(exports, "FieldType", { enumerable: true, get: function () { return types_1.FieldType; } });
