// @volcengine/openapi/lib/services/i18nOpenapi/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultService = exports.i18nOpenapiService = void 0;
const service_1 = __importDefault(require("../../base/service"));
class i18nOpenapiService extends service_1.default {
    constructor(options) {
        super(Object.assign(Object.assign({}, options), { defaultVersion: "2021-05-21", serviceName: "i18n_openapi" }));
        this.Projects = this.createAPI("Projects");
        this.ProjectDetail = this.createAPI("ProjectDetail");
        this.ProjectUsers = this.createAPI("ProjectUsers");
        this.ProjectUserRole = this.createAPI("ProjectUserRole");
        this.ProjectNamespaces = this.createAPI("ProjectNamespaces");
        this.ProjectNamespaceDetail = this.createAPI("ProjectNamespaceDetail");
        this.ProjectNamespaceCreate = this.createAPI("ProjectNamespaceCreate", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectNamespaceSources = this.createAPI("ProjectNamespaceSources");
        this.ProjectNamespaceSourceDetail = this.createAPI("ProjectNamespaceSourceDetail");
        this.ProjectNamespaceSourceDownload = this.createAPI("ProjectNamespaceSourceDownload");
        this.ProjectNamespaceSourceUpdate = this.createAPI("ProjectNamespaceSourceUpdate", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectNamespaceSourceAdd = this.createAPI("ProjectNamespaceSourceAdd", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectNamespaceSourceDeleteByKeys = this.createAPI("ProjectNamespaceSourceDeleteByKeys", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectNamespaceSourceDeleteByIds = this.createAPI("ProjectNamespaceSourceDeleteByIds", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectNamespaceTargets = this.createAPI("ProjectNamespaceTargets");
        this.ProjectNamespaceTargetDownload = this.createAPI("ProjectNamespaceTargetDownload");
        this.ProjectNamespaceTargetUpdate = this.createAPI("ProjectNamespaceTargetUpdate", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectNamespaceTargetDeleteById = this.createAPI("ProjectNamespaceTargetDeleteById", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTasks = this.createAPI("ProjectTasks");
        this.ProjectTaskDetail = this.createAPI("ProjectTaskDetail");
        this.ProjectTaskCreate = this.createAPI("ProjectTaskCreate", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTaskSources = this.createAPI("ProjectTaskSources");
        this.ProjectTaskSourceDetail = this.createAPI("ProjectTaskSourceDetail");
        this.ProjectTaskSourceDownload = this.createAPI("ProjectTaskSourceDownload");
        this.ProjectTaskSourceUpdate = this.createAPI("ProjectTaskSourceUpdate", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTaskSourceAdd = this.createAPI("ProjectTaskSourceAdd", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTaskSourceDeleteByKeys = this.createAPI("ProjectTaskSourceDeleteByKeys", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTaskSourceDeleteByIds = this.createAPI("ProjectTaskSourceDeleteByIds", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTaskTargets = this.createAPI("ProjectTaskTargets");
        this.ProjectTaskTargetUpdate = this.createAPI("ProjectTaskTargetUpdate", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTaskTargetDownload = this.createAPI("ProjectTaskTargetDownload");
        this.ProjectTaskTargetDeleteById = this.createAPI("ProjectTaskTargetDeleteById", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTerms = this.createAPI("ProjectTerms");
        this.ProjectTermDetail = this.createAPI("ProjectTermDetail");
        this.ProjectTermTargetAdd = this.createAPI("ProjectTermTargetAdd", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTermTargetUpdate = this.createAPI("ProjectTermTargetUpdate", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTermTargetDelete = this.createAPI("ProjectTermTargetDelete", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTermSourceAdd = this.createAPI("ProjectTermSourceAdd", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTermSourceUpdate = this.createAPI("ProjectTermSourceUpdate", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectTermSourceDelete = this.createAPI("ProjectTermSourceDelete", {
            method: "POST",
            contentType: "json",
        });
        this.ProjectDistributions = this.createAPI("ProjectDistributions");
        this.ProjectDistributionsRelease = this.createAPI("ProjectDistributionsRelease", {
            method: "POST",
            contentType: "json",
        });
    }
}
exports.i18nOpenapiService = i18nOpenapiService;
exports.defaultService = new i18nOpenapiService();
