// @volcengine/openapi/lib/services/cdn/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultService = exports.CdnService = void 0;
const service_1 = __importDefault(require("../../base/service"));
class CdnService extends service_1.default {
    constructor(options) {
        super(Object.assign({ host: "cdn.volcengineapi.com", serviceName: "cdn", defaultVersion: "2021-03-01" }, options));
        this.Generic = async (name, body) => {
            return await this.fetchOpenAPI({
                Action: name,
                Version: "2021-03-01",
                method: "POST",
                headers: {
                    "content-type": "application/json",
                },
                data: body,
            });
        };
        // 添加加速域名: https://www.volcengine.com/docs/6454/97340
        this.AddCdnDomain = (body) => {
            return this.Generic("AddCdnDomain", body);
        };
        // 上线加速域名: https://www.volcengine.com/docs/6454/74667
        this.StartCdnDomain = (body) => {
            return this.Generic("StartCdnDomain", body);
        };
        // 下线加速域名: https://www.volcengine.com/docs/6454/75129
        this.StopCdnDomain = (body) => {
            return this.Generic("StopCdnDomain", body);
        };
        // 删除加速域名: https://www.volcengine.com/docs/6454/75130
        this.DeleteCdnDomain = (body) => {
            return this.Generic("DeleteCdnDomain", body);
        };
        // 获取域名列表: https://www.volcengine.com/docs/6454/75269
        this.ListCdnDomains = (body) => {
            return this.Generic("ListCdnDomains", body);
        };
        // 获取域名配置详情: https://www.volcengine.com/docs/6454/80320
        this.DescribeCdnConfig = (body) => {
            return this.Generic("DescribeCdnConfig", body);
        };
        // 修改加速域名配置: https://www.volcengine.com/docs/6454/97266
        this.UpdateCdnConfig = (body) => {
            return this.Generic("UpdateCdnConfig", body);
        };
        // 获取访问统计的细分数据: https://www.volcengine.com/docs/6454/70442
        this.DescribeCdnData = (body) => {
            return this.Generic("DescribeCdnData", body);
        };
        // 获取访问统计的汇总数据: https://www.volcengine.com/docs/6454/96132
        this.DescribeEdgeNrtDataSummary = (body) => {
            return this.Generic("DescribeEdgeNrtDataSummary", body);
        };
        // 获取回源统计的细分数据: https://www.volcengine.com/docs/6454/70443
        this.DescribeCdnOriginData = (body) => {
            return this.Generic("DescribeCdnOriginData", body);
        };
        // 获取回源统计的汇总数据: https://www.volcengine.com/docs/6454/96133
        this.DescribeOriginNrtDataSummary = (body) => {
            return this.Generic("DescribeOriginNrtDataSummary", body);
        };
        // 获取省份运营商的细分数据: https://www.volcengine.com/docs/6454/75159
        this.DescribeCdnDataDetail = (body) => {
            return this.Generic("DescribeCdnDataDetail", body);
        };
        // 获取多个域名的省份和运营商的细分数据: https://www.volcengine.com/docs/6454/145577
        this.DescribeDistrictIspData = (body) => {
            return this.Generic("DescribeDistrictIspData", body);
        };
        // 获取独立访客的细分数据: https://www.volcengine.com/docs/6454/79321
        this.DescribeEdgeStatisticalData = (body) => {
            return this.Generic("DescribeEdgeStatisticalData", body);
        };
        // 获取访问统计的排行数据: https://www.volcengine.com/docs/6454/96145
        this.DescribeEdgeTopNrtData = (body) => {
            return this.Generic("DescribeEdgeTopNrtData", body);
        };
        // 获取回源数据的统计排序: https://www.volcengine.com/docs/6454/104892
        this.DescribeOriginTopNrtData = (body) => {
            return this.Generic("DescribeOriginTopNrtData", body);
        };
        // 获取访问状态码的统计排序: https://www.volcengine.com/docs/6454/104888
        this.DescribeEdgeTopStatusCode = (body) => {
            return this.Generic("DescribeEdgeTopStatusCode", body);
        };
        // 获取回源状态码的统计排序: https://www.volcengine.com/docs/6454/104891
        this.DescribeOriginTopStatusCode = (body) => {
            return this.Generic("DescribeOriginTopStatusCode", body);
        };
        // 获取热点及访客排行数据: https://www.volcengine.com/docs/6454/79322
        this.DescribeEdgeTopStatisticalData = (body) => {
            return this.Generic("DescribeEdgeTopStatisticalData", body);
        };
        // 获取区域和 ISP 列表: https://www.volcengine.com/docs/6454/70445
        this.DescribeCdnRegionAndIsp = (body) => {
            return this.Generic("DescribeCdnRegionAndIsp", body);
        };
        // 获取服务相关信息: https://www.volcengine.com/docs/6454/78999
        this.DescribeCdnService = (body) => {
            return this.Generic("DescribeCdnService", body);
        };
        // 获取计费指标的细分数据: https://www.volcengine.com/docs/6454/96167
        this.DescribeAccountingData = (body) => {
            return this.Generic("DescribeAccountingData", body);
        };
        // 提交刷新任务: https://www.volcengine.com/docs/6454/70438
        this.SubmitRefreshTask = (body) => {
            return this.Generic("SubmitRefreshTask", body);
        };
        // 提交预热任务: https://www.volcengine.com/docs/6454/70436
        this.SubmitPreloadTask = (body) => {
            return this.Generic("SubmitPreloadTask", body);
        };
        // 获取刷新预热任务信息: https://www.volcengine.com/docs/6454/70437
        this.DescribeContentTasks = (body) => {
            return this.Generic("DescribeContentTasks", body);
        };
        // 获取刷新预热配额信息: https://www.volcengine.com/docs/6454/70439
        this.DescribeContentQuota = (body) => {
            return this.Generic("DescribeContentQuota", body);
        };
        // 提交封禁任务: https://www.volcengine.com/docs/6454/79890
        this.SubmitBlockTask = (body) => {
            return this.Generic("SubmitBlockTask", body);
        };
        // 提交解封任务: https://www.volcengine.com/docs/6454/79893
        this.SubmitUnblockTask = (body) => {
            return this.Generic("SubmitUnblockTask", body);
        };
        // 获取封禁解封任务信息: https://www.volcengine.com/docs/6454/79906
        this.DescribeContentBlockTasks = (body) => {
            return this.Generic("DescribeContentBlockTasks", body);
        };
        // 获取访问日志下载链接: https://www.volcengine.com/docs/6454/70446
        this.DescribeCdnAccessLog = (body) => {
            return this.Generic("DescribeCdnAccessLog", body);
        };
        // 获取 IP 归属信息: https://www.volcengine.com/docs/6454/75233
        this.DescribeIPInfo = (body) => {
            return this.Generic("DescribeIPInfo", body);
        };
        // 批量获取 IP 归属信息: https://www.volcengine.com/docs/6454/106852
        this.DescribeIPListInfo = (body) => {
            return this.Generic("DescribeIPListInfo", body);
        };
        // 获取回源节点 IP 列表: https://www.volcengine.com/docs/6454/75273
        this.DescribeCdnUpperIp = (body) => {
            return this.Generic("DescribeCdnUpperIp", body);
        };
        // 添加资源标签: https://www.volcengine.com/docs/6454/80308
        this.AddResourceTags = (body) => {
            return this.Generic("AddResourceTags", body);
        };
        // 更新资源标签: https://www.volcengine.com/docs/6454/80313
        this.UpdateResourceTags = (body) => {
            return this.Generic("UpdateResourceTags", body);
        };
        // 查询标签清单: https://www.volcengine.com/docs/6454/80315
        this.ListResourceTags = (body) => {
            return this.Generic("ListResourceTags", body);
        };
        // 删除资源标签: https://www.volcengine.com/docs/6454/80316
        this.DeleteResourceTags = (body) => {
            return this.Generic("DeleteResourceTags", body);
        };
        // 上传证书: https://www.volcengine.com/docs/6454/125708
        this.AddCdnCertificate = (body) => {
            return this.Generic("AddCdnCertificate", body);
        };
        // 查询CDN证书列表: https://www.volcengine.com/docs/6454/125709
        this.ListCertInfo = (body) => {
            return this.Generic("ListCertInfo", body);
        };
        // 查询CDN有关联域名的证书列表: https://www.volcengine.com/docs/6454/125710
        this.ListCdnCertInfo = (body) => {
            return this.Generic("ListCdnCertInfo", body);
        };
        // 获取特定证书的域名关联信息: https://www.volcengine.com/docs/6454/125711
        this.DescribeCertConfig = (body) => {
            return this.Generic("DescribeCertConfig", body);
        };
        // 批量关联证书: https://www.volcengine.com/docs/6454/125712
        this.BatchDeployCert = (body) => {
            return this.Generic("BatchDeployCert", body);
        };
    }
}
exports.CdnService = CdnService;
exports.defaultService = new CdnService();
