// @volcengine/openapi/lib/services/tls/producer/batch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Batch = void 0;
const config_1 = require("./config");
const utils_1 = require("./utils");
const MaxLogGroupCount = 10000;
/**
 * 单个日志批次，负责在内存中聚合日志并携带重试/回调等元信息。
 */
class Batch {
    constructor(params) {
        const { topic, shardHash, config } = params;
        this.logGroupList = { LogGroups: [] };
        this.totalDataSize = 0;
        this.logCount = 0;
        this.attemptCount = 0;
        this.retryBackoffMs = 0;
        this.baseRetryBackoffMs = config.BaseRetryBackoffMs;
        this.baseIncreaseRetryBackoffMs = 1000;
        this.nextRetryMs = 0;
        this.maxRetryIntervalInMs = config.MaxRetryBackoffMs;
        this.callBackList = [];
        this.createTime = Date.now();
        this.maxRetryTimes = config.Retries;
        this.topic = topic;
        this.shardHash = shardHash;
        this.result = {
            Attempts: [],
            SuccessFlag: false,
        };
        this.maxReservedAttempts = config.MaxReservedAttempts;
    }
    /**
     * 尝试向当前批次追加一条日志。
     *
     * @param batchLog 日志内容。
     * @param callBack 该条日志对应的回调（可选）。
     * @returns 是否追加成功；若返回 false，说明需要开启新的批次。
     */
    async tryAddLog(batchLog, callBack) {
        const { Source, FileName, ContextFlow } = batchLog.Key;
        let group = this.getOrCreateTailGroup(Source, FileName, ContextFlow);
        if (group.Logs.length >= MaxLogGroupCount) {
            group = this.createGroup(Source, FileName, ContextFlow);
            this.logGroupList.LogGroups.push(group);
        }
        group.Logs.push(batchLog.Log);
        this.logCount++;
        this.totalDataSize = await (0, utils_1.getLogGroupListSize)(this.logGroupList);
        if (!this.hasRoomFor(0, 0)) {
            group.Logs.pop();
            this.logCount--;
            if (group.Logs.length === 0 &&
                this.logGroupList.LogGroups.length > 0 &&
                this.logGroupList.LogGroups[this.logGroupList.LogGroups.length - 1] === group) {
                this.logGroupList.LogGroups.pop();
            }
            return false;
        }
        if (callBack) {
            this.callBackList.push(callBack);
        }
        return true;
    }
    getOrCreateTailGroup(source, filename, contextFlow) {
        if (this.logGroupList.LogGroups.length === 0) {
            const group = this.createGroup(source, filename, contextFlow);
            this.logGroupList.LogGroups.push(group);
            return group;
        }
        const lastGroup = this.logGroupList.LogGroups[this.logGroupList.LogGroups.length - 1];
        if (lastGroup.Source === source &&
            lastGroup.FileName === filename &&
            lastGroup.ContextFlow == contextFlow) {
            return lastGroup;
        }
        const group = this.createGroup(source, filename, contextFlow);
        this.logGroupList.LogGroups.push(group);
        return group;
    }
    createGroup(source, filename, contextFlow) {
        return {
            Source: source,
            FileName: filename,
            ContextFlow: contextFlow,
            Logs: [],
            LogTags: [],
        };
    }
    /**
     * 判断按给定 size 与数量追加日志后，是否仍然满足 Batch 大小与条数限制。
     */
    hasRoomFor(size, cnt) {
        return this.totalDataSize + size <= config_1.MAX_BATCH_SIZE && this.logCount + cnt <= config_1.MAX_BATCH_COUNT;
    }
    /**
     * 判断当前批次是否已经满足发送条件：
     * - Batch 内累计大小达到或超过 MaxBatchSize；或
     * - Batch 内日志条数达到或超过 MaxBatchCount。
     */
    meetSendCondition(config) {
        return this.totalDataSize >= config.MaxBatchSize || this.logCount >= config.MaxBatchCount;
    }
}
exports.Batch = Batch;
