// @volcengine/openapi/lib/services/tls/producer/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateProducerConfig = exports.getDefaultProducerConfig = exports.TIMEOUT_EXCEPTION = exports.MAX_BATCH_COUNT = exports.MAX_BATCH_SIZE = void 0;
const utils_1 = require("../utils");
/**
 * 单个 Batch 的最大字节数硬上限，单位：字节。
 */
exports.MAX_BATCH_SIZE = 8 * 1024 * 1024; // 8 MB
/**
 * 单个 Batch 的最大日志条数硬上限。
 */
exports.MAX_BATCH_COUNT = 32768;
/** TimeoutException 文本 */
exports.TIMEOUT_EXCEPTION = "TimeoutException";
/**
 * 根据环境变量和默认值生成一份 Producer 配置。
 */
function getDefaultProducerConfig() {
    const defaults = (0, utils_1.getDefaultOption)();
    return {
        TotalSizeLnBytes: 100 * 1024 * 1024,
        MaxSenderCount: 50,
        MaxBlockSec: 60,
        MaxBatchSize: 512 * 1024,
        MaxBatchCount: 4096,
        LingerTime: 2000,
        Retries: 10,
        MaxReservedAttempts: 11,
        BaseRetryBackoffMs: 1000,
        MaxRetryBackoffMs: 10 * 1000,
        AdjustShardHashFlag: true,
        ShardCount: 2,
        NoRetryStatusCodeList: [400, 404],
        Endpoint: defaults.host,
        Region: defaults.region,
        AccessKeyID: defaults.accessKeyId,
        AccessKeySecret: defaults.secretKey,
        SecurityToken: "",
    };
}
exports.getDefaultProducerConfig = getDefaultProducerConfig;
function validateNumberField(val, min, max, defaultVal) {
    if (typeof val !== "number" || Number.isNaN(val))
        return defaultVal;
    if (val < min || val > max)
        return defaultVal;
    return val;
}
/**
 * 将用户传入的配置与默认值合并，并做边界校验与修正。
 */
function validateProducerConfig(config) {
    const base = getDefaultProducerConfig();
    const merged = Object.assign(Object.assign({}, base), config);
    merged.MaxReservedAttempts = validateNumberField(merged.MaxReservedAttempts, 0, Number.MAX_SAFE_INTEGER, base.MaxReservedAttempts);
    merged.MaxBatchCount = validateNumberField(merged.MaxBatchCount, 0, exports.MAX_BATCH_COUNT, base.MaxBatchCount);
    merged.MaxBatchSize = validateNumberField(merged.MaxBatchSize, 0, exports.MAX_BATCH_SIZE, base.MaxBatchSize);
    merged.MaxSenderCount = validateNumberField(merged.MaxSenderCount, 0, Number.MAX_SAFE_INTEGER, base.MaxSenderCount);
    merged.BaseRetryBackoffMs = validateNumberField(merged.BaseRetryBackoffMs, 0, Number.MAX_SAFE_INTEGER, base.BaseRetryBackoffMs);
    merged.MaxRetryBackoffMs = validateNumberField(merged.MaxRetryBackoffMs, 0, Number.MAX_SAFE_INTEGER, base.MaxRetryBackoffMs);
    merged.TotalSizeLnBytes = validateNumberField(merged.TotalSizeLnBytes, 0, Number.MAX_SAFE_INTEGER, base.TotalSizeLnBytes);
    merged.LingerTime = validateNumberField(merged.LingerTime, 100, Number.MAX_SAFE_INTEGER, base.LingerTime);
    if (!Array.isArray(merged.NoRetryStatusCodeList) || merged.NoRetryStatusCodeList.length === 0) {
        merged.NoRetryStatusCodeList = [400, 404];
    }
    return merged;
}
exports.validateProducerConfig = validateProducerConfig;
