// @volcengine/openapi/lib/services/tls/producer/sender.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sender = void 0;
const service_1 = __importDefault(require("../service"));
const utils_1 = require("./utils");
const utils_2 = require("./utils");
const logger_1 = __importDefault(require("../logger"));
/**
 * 负责将 Batch 发送到 TLS 服务，并在失败时按照重试策略将批次重新放入重试队列。
 */
class Sender {
    constructor(params) {
        this.stopped = false;
        const { service, retryQueue, config, headers, noRetryStatusCodeList, decreaseBufferedSize } = params;
        this.service = service;
        this.retryQueue = retryQueue;
        this.config = config;
        this.headers = headers;
        this.noRetryStatusCodeSet = new Set(noRetryStatusCodeList || []);
        this.decreaseBufferedSize = decreaseBufferedSize;
        this.putLogs = this.service.createPutLogsAPI("PutLogs");
    }
    stop() {
        this.stopped = true;
    }
    /**
     * 将一个批次发送到服务端；若失败则根据状态码和重试次数决定是否进入重试队列。
     */
    async send(batch) {
        logger_1.default.debug("sending batch data to server");
        if (this.stopped)
            return;
        // 使用 objToProtoBuffer 序列化为 protobuf Buffer
        const pbMessage = await service_1.default.objToProtoBuffer(batch.logGroupList);
        const req = {
            TopicId: batch.topic,
            CompressType: "lz4",
            LogGroupList: Buffer.from(pbMessage),
        };
        if (batch.shardHash) {
            req.HashKey = batch.shardHash;
        }
        try {
            const res = await this.putLogs(req, this.headers ? { headers: this.headers } : undefined);
            this.handleSuccess(batch, res);
        }
        catch (err) {
            this.handleFailure(batch, err);
        }
    }
    handleSuccess(batch, res) {
        logger_1.default.debug("batch data sent successfully to server");
        // 发送成功，减少 Producer 缓冲区占用
        this.decreaseBufferedSize(batch.totalDataSize);
        batch.result.SuccessFlag = true;
        if (batch.attemptCount < batch.maxReservedAttempts) {
            const attempt = {
                SuccessFlag: true,
                RequestId: (res === null || res === void 0 ? void 0 : res.RequestID) || "",
                ErrorCode: "",
                ErrorMessage: "",
                TimestampMs: (0, utils_1.getTimeMs)(),
            };
            batch.result.Attempts.push(attempt);
        }
        // 执行成功回调
        for (const cb of batch.callBackList) {
            (0, utils_2.withRecover)(() => cb.Success(batch.result));
        }
        batch.retryBackoffMs = 0;
    }
    handleFailure(batch, error) {
        logger_1.default.error("batch data sent failed to server", { error });
        const { noRetryStatusCodeSet, config } = this;
        const { statusCode, requestId, errorCode, errorMessage } = this.extractErrorInfo(error);
        const noRetryStatusCode = typeof statusCode === "number" && noRetryStatusCodeSet.has(statusCode);
        const noNeedRetry = batch.attemptCount >= batch.maxRetryTimes;
        if (this.stopped || noRetryStatusCode || noNeedRetry) {
            this.addErrorAttempt(batch, {
                requestId,
                errorCode,
                errorMessage,
            });
            this.failedCallback(batch);
            return;
        }
        // 满足重试条件，先记录一次失败尝试，再放入重试队列
        this.addErrorAttempt(batch, {
            requestId,
            errorCode,
            errorMessage,
        });
        if (batch.attemptCount === 1) {
            batch.retryBackoffMs += config.BaseRetryBackoffMs;
        }
        else {
            const increase = Math.random() * batch.baseIncreaseRetryBackoffMs;
            batch.retryBackoffMs += Math.floor(increase);
        }
        if (batch.retryBackoffMs > config.MaxRetryBackoffMs) {
            batch.retryBackoffMs = config.MaxRetryBackoffMs;
        }
        batch.nextRetryMs = (0, utils_1.getTimeMs)() + batch.retryBackoffMs;
        this.retryQueue.add(batch);
    }
    failedCallback(batch) {
        // 最终失败同样需要释放内存占用
        this.decreaseBufferedSize(batch.totalDataSize);
        for (const cb of batch.callBackList) {
            (0, utils_2.withRecover)(() => cb.Fail(batch.result));
        }
    }
    addErrorAttempt(batch, info) {
        if (batch.attemptCount < batch.maxReservedAttempts) {
            const attempt = {
                SuccessFlag: false,
                RequestId: info.requestId || "",
                ErrorCode: info.errorCode || "",
                ErrorMessage: info.errorMessage || "",
                TimestampMs: (0, utils_1.getTimeMs)(),
            };
            batch.result.Attempts.push(attempt);
        }
        batch.result.SuccessFlag = false;
        batch.attemptCount += 1;
    }
    /**
     * 从 AxiosError 中解析状态码、RequestId、错误码与错误信息。
     */
    extractErrorInfo(error) {
        const resp = error.response;
        const statusCode = resp === null || resp === void 0 ? void 0 : resp.status;
        let requestId;
        let errorCode;
        let errorMessage;
        const headers = (resp === null || resp === void 0 ? void 0 : resp.headers) || {};
        const lowerHeaders = {};
        Object.keys(headers).forEach((key) => {
            const val = headers[key];
            if (typeof val === "string") {
                lowerHeaders[key.toLowerCase()] = val;
            }
        });
        requestId =
            lowerHeaders["x-tls-requestid"] ||
                lowerHeaders["x-tls-request-id"] ||
                lowerHeaders["x-request-id"];
        const data = resp === null || resp === void 0 ? void 0 : resp.data;
        if (data) {
            if (typeof data === "string") {
                errorMessage = data;
            }
            else if (typeof data === "object") {
                if (data.Error) {
                    errorCode = data.Error.Code || data.Error.code;
                    errorMessage = data.Error.Message || data.Error.message;
                    requestId = requestId || data.RequestId || data.requestId || data.RequestID;
                }
                else {
                    errorCode = data.Code || data.code;
                    errorMessage = data.Message || data.message;
                    requestId = requestId || data.RequestId || data.requestId || data.RequestID;
                }
            }
        }
        if (!errorMessage) {
            errorMessage = error.message;
        }
        return { statusCode, requestId, errorCode, errorMessage };
    }
}
exports.Sender = Sender;
