// @volcengine/openapi/lib/services/tls/producer/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.withRecover = exports.sleep = exports.getTimeMs = exports.getLogGroupListSize = exports.getLogGroupSize = exports.getLogSize = void 0;
const protobufjs_1 = __importDefault(require("protobufjs"));
const path_1 = __importDefault(require("path"));
let logTypePromise = null;
let logGroupTypePromise = null;
let logGroupListTypePromise = null;
function getProtoFilePath() {
    // 运行时位于 lib/services/tls/producer，下探一级即可找到拷贝过来的 tls.proto
    return path_1.default.join(__dirname, "..", "tls.proto");
}
async function getLogType() {
    if (!logTypePromise) {
        logTypePromise = protobufjs_1.default
            .load(getProtoFilePath())
            .then((root) => root.lookupType("pb.Log"));
    }
    return logTypePromise;
}
async function getLogGroupType() {
    if (!logGroupTypePromise) {
        logGroupTypePromise = protobufjs_1.default
            .load(getProtoFilePath())
            .then((root) => root.lookupType("pb.LogGroup"));
    }
    return logGroupTypePromise;
}
async function getLogGroupListType() {
    if (!logGroupListTypePromise) {
        logGroupListTypePromise = protobufjs_1.default
            .load(getProtoFilePath())
            .then((root) => root.lookupType("pb.LogGroupList"));
    }
    return logGroupListTypePromise;
}
/**
 * 计算单条 Log 的 protobuf 编码字节数，语义接近 Go 中的 pb.Log.Size()。
 */
async function getLogSize(log) {
    const type = await getLogType();
    const errMsg = type.verify(log);
    if (errMsg) {
        throw new Error(`[tls-node-sdk] invalid Log: ${errMsg}`);
    }
    const message = type.create(log);
    return type.encode(message).finish().length;
}
exports.getLogSize = getLogSize;
/**
 * 计算 LogGroup 的 protobuf 编码字节数，便于精确预估整个批次大小。
 */
async function getLogGroupSize(logGroup) {
    const type = await getLogGroupType();
    const errMsg = type.verify(logGroup);
    if (errMsg) {
        throw new Error(`[tls-node-sdk] invalid LogGroup: ${errMsg}`);
    }
    const message = type.create(logGroup);
    return type.encode(message).finish().length;
}
exports.getLogGroupSize = getLogGroupSize;
async function getLogGroupListSize(logGroupList) {
    const type = await getLogGroupListType();
    const errMsg = type.verify(logGroupList);
    if (errMsg) {
        throw new Error(`[tls-node-sdk] invalid LogGroupList: ${errMsg}`);
    }
    const message = type.create(logGroupList);
    return type.encode(message).finish().length;
}
exports.getLogGroupListSize = getLogGroupListSize;
/**
 * 获取当前时间的毫秒时间戳。
 */
function getTimeMs() {
    return Date.now();
}
exports.getTimeMs = getTimeMs;
/**
 * 简单的异步 sleep 工具，单位：毫秒。
 */
function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
exports.sleep = sleep;
/**
 * 封装 panic/recover 语义的工具函数，避免回调抛出的异常影响主流程。
 */
function withRecover(fn) {
    try {
        fn();
    }
    catch (err) {
        // 在生产环境中可以按需接入统一日志组件。
        // eslint-disable-next-line no-console
        console.error(err);
    }
}
exports.withRecover = withRecover;
