// @volcengine/openapi/lib/services/tls/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultService = exports.TlsService = void 0;
const service_1 = __importDefault(require("./service"));
const SERVICE = "TLS";
// const REGION = "cn-beijing";
const VERSION = "0.2.0";
const PROTOCOL = "https:";
class TlsService extends service_1.default {
    constructor(options) {
        super(Object.assign({ version: VERSION, serviceName: SERVICE, protocol: PROTOCOL }, options));
        this.DescribeProjects = this.createAPI("DescribeProjects", {
            method: "GET",
        });
        this.DescribeProject = this.createAPI("DescribeProject", {
            method: "GET",
        });
        this.CreateProject = this.createAPI("CreateProject", {
            method: "POST",
        });
        this.ModifyProject = this.createAPI("ModifyProject", {
            method: "PUT",
        });
        this.DeleteProject = this.createAPI("DeleteProject", {
            method: "DELETE",
        });
        // log topic related interface
        // 日志主题相关接口
        this.DescribeTopics = this.createAPI("DescribeTopics", {
            method: "GET",
        });
        this.DescribeTopic = this.createAPI("DescribeTopic", {
            method: "GET",
        });
        this.CreateTopic = this.createAPI("CreateTopic", {
            method: "POST",
        });
        this.ModifyTopic = this.createAPI("ModifyTopic", {
            method: "PUT",
        });
        this.DeleteTopic = this.createAPI("DeleteTopic", {
            method: "DELETE",
        });
        // index related interface
        // 索引相关接口
        this.DescribeIndex = this.createAPI("DescribeIndex", {
            method: "GET",
        });
        this.CreateIndex = this.createAPI("CreateIndex", {
            method: "POST",
        });
        this.ModifyIndex = this.createAPI("ModifyIndex", {
            method: "PUT",
        });
        this.DeleteIndex = this.createAPI("DeleteIndex", {
            method: "DELETE",
        });
        // topic partition related interface
        // 主题分区相关接口
        this.DescribeShards = this.createAPI("DescribeShards", {
            method: "GET",
        });
        this.ManualShardSplit = this.createAPI("ManualShardSplit", {
            method: "POST",
        });
        this.DescribeCursor = this.createAPI("DescribeCursor", {
            method: "GET",
            queryKeys: ["TopicId", "ShardId"],
        });
        // collection and configuration related interfaces
        // 采集配置相关接口
        this.CreateRule = this.createAPI("CreateRule", {
            method: "POST",
        });
        this.DeleteRule = this.createAPI("DeleteRule", {
            method: "DELETE",
        });
        this.ModifyRule = this.createAPI("ModifyRule", {
            method: "PUT",
        });
        this.DescribeRule = this.createAPI("DescribeRule", {
            method: "GET",
        });
        this.DescribeRules = this.createAPI("DescribeRules", {
            method: "GET",
        });
        // collect and configure binding/unbinding host groups
        // 采集配置绑定/解绑机器组
        this.ApplyRuleToHostGroups = this.createAPI("ApplyRuleToHostGroups", {
            method: "PUT",
        });
        this.DeleteRuleFromHostGroups = this.createAPI("DeleteRuleFromHostGroups", {
            method: "PUT",
        });
        // host group
        // 机器组
        this.CreateHostGroup = this.createAPI("CreateHostGroup", {
            method: "POST",
        });
        this.DeleteHostGroup = this.createAPI("DeleteHostGroup", {
            method: "DELETE",
        });
        this.ModifyHostGroup = this.createAPI("ModifyHostGroup", {
            method: "PUT",
        });
        this.ModifyHostGroupsAutoUpdate = this.createAPI("ModifyHostGroupsAutoUpdate", {
            method: "PUT",
        });
        this.DescribeHostGroup = this.createAPI("DescribeHostGroup", {
            method: "GET",
        });
        this.DescribeHostGroups = this.createAPI("DescribeHostGroups", {
            method: "GET",
        });
        this.DescribeHosts = this.createAPI("DescribeHosts", {
            method: "GET",
        });
        this.DeleteHost = this.createAPI("DeleteHost", {
            method: "DELETE",
        });
        this.DeleteAbnormalHosts = this.createAPI("DeleteAbnormalHosts", {
            method: "DELETE",
        });
        this.DescribeHostGroupRules = this.createAPI("DescribeHostGroupRules", {
            method: "GET",
        });
        // alarm
        // 告警
        this.CreateAlarm = this.createAPI("CreateAlarm", {
            method: "POST",
        });
        this.DeleteAlarm = this.createAPI("DeleteAlarm", {
            method: "DELETE",
        });
        this.DescribeAlarms = this.createAPI("DescribeAlarms", {
            method: "GET",
        });
        this.ModifyAlarm = this.createAPI("ModifyAlarm", {
            method: "PUT",
        });
        this.CreateAlarmNotifyGroup = this.createAPI("CreateAlarmNotifyGroup", {
            method: "POST",
        });
        this.DeleteAlarmNotifyGroup = this.createAPI("DeleteAlarmNotifyGroup", {
            method: "DELETE",
        });
        this.ModifyAlarmNotifyGroup = this.createAPI("ModifyAlarmNotifyGroup", {
            method: "PUT",
        });
        this.DescribeAlarmNotifyGroups = this.createAPI("DescribeAlarmNotifyGroups", {
            method: "GET",
        });
        this.PutLogs = this.createPutLogsAPI("PutLogs");
        this.SearchLogs = this.createAPI("SearchLogs", {
            method: "POST",
        });
        this.DescribeHistogramV1 = this.createAPI("DescribeHistogramV1", {
            method: "POST",
        });
        this.ConsumeLogs = this.createAPI("ConsumeLogs", {
            method: "GET",
            queryKeys: ["TopicId", "ShardId"],
            axiosConfig: {
                responseType: "arraybuffer",
                headers: {
                    Accept: "application/x-protobuf",
                },
            },
        });
        this.DescribeLogContext = this.createAPI("DescribeLogContext", {
            method: "POST",
        });
        this.DescribeTraceInstances = this.createAPI("DescribeTraceInstances", {
            method: "GET",
        });
        this.DeleteTraceInstance = this.createAPI("DeleteTraceInstance", {
            method: "delete",
        });
        this.CreateTraceInstance = this.createAPI("CreateTraceInstance", {
            method: "POST",
        });
        this.ModifyTraceInstance = this.createAPI("ModifyTraceInstance", {
            method: "PUT",
        });
        this.DescribeTraceInstance = this.createAPI("DescribeTraceInstance", {
            method: "GET",
        });
        this.DescribeShipper = this.createAPI("DescribeShipper", {
            method: "GET",
        });
        this.DescribeShippers = this.createAPI("DescribeShippers", {
            method: "GET",
        });
        this.DescribeTrace = this.createAPI("DescribeTrace", {
            method: "POST",
        });
        this.SearchTraces = this.createAPI("SearchTraces", {
            method: "POST",
        });
        this.DescribeETLTask = this.createAPI("DescribeETLTask", {
            method: "GET",
        });
        this.DescribeETLTasks = this.createAPI("DescribeETLTasks", {
            method: "GET",
        });
        this.DescribeImportTask = this.createAPI("DescribeImportTask", {
            method: "GET",
        });
        this.GetAccountStatus = this.createAPI("GetAccountStatus", {
            method: "GET",
        });
        this.ActiveTlsAccount = this.createAPI("ActiveTlsAccount", {
            method: "POST",
        });
        this.CreateImportTask = this.createAPI("CreateImportTask", {
            method: "POST",
        });
        // consumer group related interface
        // 消费组相关接口
        this.CreateConsumerGroup = this.createAPI("CreateConsumerGroup", {
            method: "POST",
        });
        this.DeleteShipper = this.createAPI("DeleteShipper", {
            method: "DELETE",
        });
        this.DescribeConsumerGroups = this.createAPI("DescribeConsumerGroups", {
            method: "GET",
        });
        this.ModifyConsumerGroup = this.createAPI("ModifyConsumerGroup", {
            method: "PUT",
        });
        this.DeleteConsumerGroup = this.createAPI("DeleteConsumerGroup", {
            method: "DELETE",
        });
        this.DescribeCheckPoint = this.createAPI("DescribeCheckPoint", {
            method: "GET",
            queryKeys: ["ProjectId", "TopicId", "ShardId"],
        });
        this.ModifyCheckPoint = this.createAPI("ModifyCheckPoint", {
            method: "PUT",
        });
        this.ResetCheckPoint = this.createAPI("ResetCheckPoint", {
            method: "PUT",
        });
        this.ListTagsForResources = this.createAPI("ListTagsForResources", {
            method: "POST",
        });
        this.AddTagsToResource = this.createAPI("AddTagsToResource", {
            method: "POST",
        });
        this.TagResources = this.createAPI("TagResources", {
            method: "POST",
        });
        this.UntagResources = this.createAPI("UntagResources", {
            method: "POST",
        });
        this.RemoveTagsFromResource = this.createAPI("RemoveTagsFromResource", {
            method: "POST",
        });
        this.CancelDownloadTask = this.createAPI("CancelDownloadTask", {
            method: "POST",
        });
        this.ModifyShipper = this.createAPI("ModifyShipper", {
            method: "PUT",
        });
        this.CloseKafkaConsumer = this.createAPI("CloseKafkaConsumer", {
            method: "PUT",
        });
        this.OpenKafkaConsumer = this.createAPI("OpenKafkaConsumer", {
            method: "PUT",
        });
        this.DescribeKafkaConsumer = this.createAPI("DescribeKafkaConsumer", {
            method: "GET",
        });
        this.ConsumerHeartbeat = this.createAPI("ConsumerHeartbeat", {
            method: "POST",
        });
        this.ModifyETLTaskStatus = this.createAPI("ModifyETLTaskStatus", {
            method: "PUT",
        });
        this.CreateDownloadTask = this.createAPI("CreateDownloadTask", {
            method: "POST",
        });
        this.DescribeDownloadTasks = this.createAPI("DescribeDownloadTasks", {
            method: "GET",
        });
        this.DescribeDownloadUrl = this.createAPI("DescribeDownloadUrl", {
            method: "GET",
        });
        this.WebTracks = this.createAPI("WebTracks", {
            method: "POST",
            queryKeys: ["ProjectId", "TopicId"],
        });
        this.ModifyETLTask = this.createAPI("ModifyETLTask", {
            method: "PUT",
        });
        this.CreateETLTask = this.createAPI("CreateETLTask", {
            method: "POST",
        });
        this.DeleteETLTask = this.createAPI("DeleteETLTask", {
            method: "DELETE",
        });
        this.DeleteImportTask = this.createAPI("DeleteImportTask", {
            method: "DELETE",
        });
        this.CreateShipper = this.createAPI("CreateShipper", {
            method: "POST",
        });
        this.CreateScheduleSqlTask = this.createAPI("CreateScheduleSqlTask", {
            method: "POST",
        });
        this.DeleteScheduleSqlTask = this.createAPI("DeleteScheduleSqlTask", {
            method: "DELETE",
        });
        this.DescribeScheduleSqlTask = this.createAPI("DescribeScheduleSqlTask", {
            method: "GET",
        });
        this.DescribeScheduleSqlTasks = this.createAPI("DescribeScheduleSqlTasks", {
            method: "GET",
        });
        this.ModifyScheduleSqlTask = this.createAPI("ModifyScheduleSqlTask", {
            method: "PUT",
        });
        this.CreateAlarmContentTemplate = this.createAPI("CreateAlarmContentTemplate", {
            method: "POST",
        });
        this.DeleteAlarmContentTemplate = this.createAPI("DeleteAlarmContentTemplate", {
            method: "DELETE",
        });
        this.ModifyAlarmContentTemplate = this.createAPI("ModifyAlarmContentTemplate", {
            method: "PUT",
        });
        this.DescribeAlarmContentTemplates = this.createAPI("DescribeAlarmContentTemplates", {
            method: "GET",
        });
        this.CreateAlarmWebhookIntegration = this.createAPI("CreateAlarmWebhookIntegration", {
            method: "POST",
        });
        this.DeleteAlarmWebhookIntegration = this.createAPI("DeleteAlarmWebhookIntegration", {
            method: "DELETE",
        });
        this.ModifyAlarmWebhookIntegration = this.createAPI("ModifyAlarmWebhookIntegration", {
            method: "PUT",
        });
        this.DescribeAlarmWebhookIntegrations = this.createAPI("DescribeAlarmWebhookIntegrations", {
            method: "GET",
        });
    }
}
exports.TlsService = TlsService;
exports.defaultService = new TlsService();
__exportStar(require("./producer"), exports);
