// @volcengine/openapi/lib/services/tls/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = exports.LogLevel = void 0;
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["Debug"] = 0] = "Debug";
    LogLevel[LogLevel["Info"] = 1] = "Info";
    LogLevel[LogLevel["Warn"] = 2] = "Warn";
    LogLevel[LogLevel["Error"] = 3] = "Error";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
const DEFAULT_LOGGER_CONFIG = {
    LogLevel: LogLevel.Info,
    IsJsonType: false,
};
/**
 * logfmt 格式中单个 key=value 片段的格式化。
 *
 * 规则说明：
 * - 对于字符串，如果包含空格或等号，使用 JSON 字符串（带引号）包裹；
 * - 对于其他类型，直接使用 String(value)；
 * - 最终拼接为 `key=value` 形式。
 */
function formatLogfmtPair(key, value) {
    let rendered;
    if (typeof value === "string") {
        if (/\s|=/.test(value)) {
            rendered = JSON.stringify(value);
        }
        else {
            rendered = value;
        }
    }
    else {
        rendered = String(value);
    }
    return `${key}=${rendered}`;
}
/**
 * 将日志对象转换为 logfmt 字符串。
 */
function toLogfmt(record) {
    const parts = [];
    for (const key of Object.keys(record)) {
        const value = record[key];
        parts.push(formatLogfmtPair(key, value));
    }
    return parts.join(" ");
}
class Logger {
    constructor(config) {
        this.logger = console;
        const merged = Object.assign(Object.assign({}, DEFAULT_LOGGER_CONFIG), (config || {}));
        this.currentLevel = merged.LogLevel || LogLevel.Info;
        this.config = merged;
    }
    debug(msg, fields) {
        this.log(LogLevel.Debug, msg, fields);
    }
    info(msg, fields) {
        this.log(LogLevel.Info, msg, fields);
    }
    warn(msg, fields) {
        this.log(LogLevel.Warn, msg, fields);
    }
    error(msg, fields) {
        this.log(LogLevel.Error, msg, fields);
    }
    /**
     * 内部统一日志输出实现：根据级别做过滤，并选择 JSON 或 logfmt 格式打印。
     */
    log(level, msg, fields) {
        // 级别过滤：当前日志级别数值小于配置级别则丢弃。
        if (level < this.currentLevel) {
            return;
        }
        const now = new Date().toISOString();
        // 基础字段：time / level / msg
        const record = {
            time: now,
            level,
            msg,
        };
        // 附加字段：用户传入的键值对，避免覆盖基础字段。
        if (fields) {
            for (const key of Object.keys(fields)) {
                if (key === "time" || key === "level" || key === "msg") {
                    continue;
                }
                record[key] = fields[key];
            }
        }
        if (this.config.IsJsonType) {
            this.printJson(record, level);
        }
        else {
            this.printLogfmt(record, level);
        }
    }
    /**
     * JSON 输出：序列化为单行 JSON 字符串。
     */
    printJson(record, level) {
        const line = JSON.stringify(record);
        switch (level) {
            case LogLevel.Debug:
                this.logger.debug(line);
                break;
            case LogLevel.Info:
                this.logger.info(line);
                break;
            case LogLevel.Warn:
                this.logger.warn(line);
                break;
            case LogLevel.Error:
            default:
                this.logger.error(line);
                break;
        }
    }
    /**
     * logfmt 输出：key=value key2=value2 ...
     */
    printLogfmt(record, level) {
        const line = toLogfmt(record);
        switch (level) {
            case LogLevel.Debug:
                this.logger.debug(line);
                break;
            case LogLevel.Info:
                this.logger.info(line);
                break;
            case LogLevel.Warn:
                this.logger.warn(line);
                break;
            case LogLevel.Error:
            default:
                this.logger.error(line);
                break;
        }
    }
}
exports.Logger = Logger;
exports.default = new Logger();
