// @volcengine/openapi/lib/services/tls/service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const sign_1 = __importDefault(require("../../base/sign"));
const protobufjs_1 = __importDefault(require("protobufjs"));
const lz4_1 = __importDefault(require("./lz4"));
const zlib_1 = __importDefault(require("zlib"));
const crypto_1 = __importDefault(require("crypto"));
const path_1 = __importDefault(require("path"));
const utils_1 = require("./utils");
const TIMEOUT = 100000;
const ANONYMOUS_IDENTITY_HEADER = "x-tls-anonymous-identity";
function hasNonEmptyHeader(headers, headerName) {
    if (!headers || typeof headers !== "object")
        return false;
    for (const [key, value] of Object.entries(headers)) {
        if (key.toLowerCase() !== headerName.toLowerCase())
            continue;
        if (typeof value === "string" && value.trim())
            return true;
        if (Array.isArray(value) && value.some((v) => typeof v === "string" && v.trim()))
            return true;
    }
    return false;
}
const defaultOptions = (0, utils_1.getDefaultOption)();
class Service {
    constructor(options) {
        this.setRegion = (region) => {
            this.options.region = region;
        };
        this.setHost = (host) => (this.options.host = host);
        this.setVersion = (version) => (this.options.version = version);
        this.setProtocol = (protocol) => (this.options.protocol = protocol);
        this.setAccessKeyId = (accessKeyId) => (this.options.accessKeyId = accessKeyId);
        this.setSecretKey = (secretKey) => (this.options.secretKey = secretKey);
        this.setSessionToken = (sessionToken) => (this.options.sessionToken = sessionToken);
        this.options = Object.assign(Object.assign({}, defaultOptions), options);
    }
    static async fetch(uri, requestObj) {
        const res = await (0, axios_1.default)(Object.assign({ url: uri, timeout: TIMEOUT }, requestObj));
        return res === null || res === void 0 ? void 0 : res.data;
    }
    /**
     * transfer json to protobuf buffer
     */
    static async objToProtoBuffer(obj) {
        const root = await protobufjs_1.default.load(path_1.default.join(__dirname, "./tls.proto"));
        const type = root.lookupType("pb.LogGroupList");
        const errMsg = type.verify(obj);
        if (errMsg)
            throw new Error(errMsg);
        const message = type.create(obj);
        return type.encode(message).finish();
    }
    // protobuf buffer compress
    static LZ4Compress(input) {
        let output = Buffer.alloc(lz4_1.default.encodeBound(input.length));
        const compressedSize = lz4_1.default.encodeBlock(input, output);
        if (!compressedSize) {
            throw new Error("no need to compress");
        }
        output = output.slice(0, compressedSize);
        return output;
    }
    createAPI(Path, createParams) {
        const { method = "GET", version: versionFromParam, queryKeys = [], axiosConfig: axiosConfigFromCreateParams, } = createParams || {};
        const queryKeySet = queryKeys.length ? new Set(queryKeys) : undefined;
        return async (requestData, config) => {
            const { accessKeyId, secretKey, sessionToken, host, protocol, version: versionFromOptions, serviceName, region, } = this.options;
            const version = versionFromParam || versionFromOptions;
            const anonymousIdentityEnabled = hasNonEmptyHeader(axiosConfigFromCreateParams === null || axiosConfigFromCreateParams === void 0 ? void 0 : axiosConfigFromCreateParams.headers, ANONYMOUS_IDENTITY_HEADER) ||
                hasNonEmptyHeader(config === null || config === void 0 ? void 0 : config.headers, ANONYMOUS_IDENTITY_HEADER);
            if (!host) {
                throw new Error(`[tls-node-sdk] host is necessary`);
            }
            if (!anonymousIdentityEnabled && (!accessKeyId || !secretKey)) {
                throw new Error(`[tls-node-sdk] accessKeyId and secretKey is necessary`);
            }
            const requestObj = Object.assign(Object.assign(Object.assign(Object.assign({}, (axiosConfigFromCreateParams !== null && axiosConfigFromCreateParams !== void 0 ? axiosConfigFromCreateParams : {})), { region,
                method, pathname: `/${Path}` }), config), { headers: Object.assign(Object.assign(Object.assign({}, axiosConfigFromCreateParams === null || axiosConfigFromCreateParams === void 0 ? void 0 : axiosConfigFromCreateParams.headers), config === null || config === void 0 ? void 0 : config.headers), { "x-tls-apiversion": version }) });
            if (queryKeys.length > 0) {
                const query = {};
                const body = {};
                for (const [key, val] of Object.entries(requestData)) {
                    if (queryKeySet === null || queryKeySet === void 0 ? void 0 : queryKeySet.has(key)) {
                        if (val !== undefined)
                            query[key] = val;
                        continue;
                    }
                    body[key] = val;
                }
                requestObj.params = query;
                requestObj.body = body;
                requestObj.data = body;
                if (method === "POST") {
                    requestObj.headers["content-type"] = "application/json";
                }
            }
            else if (method === "GET") {
                requestObj.params = requestData;
            }
            else {
                requestObj.headers["content-type"] = "application/json";
                requestObj.body = requestData;
                requestObj.data = requestData;
            }
            if (!anonymousIdentityEnabled) {
                const signer = new sign_1.default(requestObj, serviceName);
                signer.addAuthorization({
                    accessKeyId: accessKeyId,
                    secretKey: secretKey,
                    sessionToken,
                });
            }
            return Service.fetch(`${protocol}//${host}/${Path}`.trim(), requestObj);
        };
    }
    createPutLogsAPI(Path) {
        return async (requestData, config) => {
            const { LogGroupList, CompressType, TopicId, HashKey, ContentMD5 } = requestData;
            if (!LogGroupList)
                throw new Error(`LogGroupList is necessary`);
            // transfer json to compressed protobuf buffer
            const pbMessage = LogGroupList;
            let compressFailed = false;
            let output = pbMessage;
            if (CompressType === "lz4") {
                try {
                    output = Service.LZ4Compress(pbMessage);
                }
                catch (err) {
                    compressFailed = true;
                }
            }
            else if (CompressType === "zlib") {
                try {
                    output = zlib_1.default.deflateSync(pbMessage);
                }
                catch (err) {
                    compressFailed = true;
                }
            }
            const { accessKeyId, secretKey, sessionToken, host, version, protocol, serviceName, region } = this.options;
            const anonymousIdentityEnabled = hasNonEmptyHeader(config === null || config === void 0 ? void 0 : config.headers, ANONYMOUS_IDENTITY_HEADER);
            if (!host || !region) {
                const missingParams = [];
                if (!host)
                    missingParams.push("host");
                if (!region)
                    missingParams.push("region");
                throw new Error(`[tls-node-sdk] ${missingParams.join(" and ")} is necessary`);
            }
            if (!anonymousIdentityEnabled && (!accessKeyId || !secretKey)) {
                const missingParams = [];
                if (!accessKeyId)
                    missingParams.push("accessKeyId");
                if (!secretKey)
                    missingParams.push("secretKey");
                throw new Error(`[tls-node-sdk] ${missingParams.join(" and ")} is necessary`);
            }
            const bodyRawSize = pbMessage.length;
            const headers = Object.assign(Object.assign({}, config === null || config === void 0 ? void 0 : config.headers), { "x-tls-apiversion": version, "Content-Type": "application/x-protobuf", "x-tls-bodyrawsize": String(bodyRawSize), "Content-MD5": ContentMD5 || crypto_1.default.createHash("md5").update(output).digest("hex") });
            if (!compressFailed && CompressType) {
                headers["x-tls-compresstype"] = CompressType;
            }
            if (HashKey && typeof HashKey === "string") {
                const normalizedHashKey = HashKey.trim();
                if (/^[0-9a-fA-F]{32}$/.test(normalizedHashKey)) {
                    headers["x-tls-hashkey"] = normalizedHashKey;
                }
            }
            const requestObj = Object.assign(Object.assign({ region, method: "POST", pathname: `/${Path}` }, config), { headers });
            requestObj.body = output;
            requestObj.data = output;
            requestObj.params = {
                TopicId,
            };
            if (!anonymousIdentityEnabled) {
                const signer = new sign_1.default(requestObj, serviceName);
                signer.addAuthorization({
                    accessKeyId: accessKeyId,
                    secretKey: secretKey,
                    sessionToken,
                });
            }
            return Service.fetch(`${protocol}//${host}/${Path}`.trim(), requestObj);
        };
    }
}
exports.default = Service;
